## Snakefile for analyses into the imputation and phasing of low cov sequencing data in Anopheles gambiae
import pandas as pd
import numpy as np

dataset = 'ag3_gaardian'
chroms = ['2L', '2R', '3R', '3L', 'X']


rule all:
    input:
        expand("resources/vcfs/{dataset}_{chrom}{allelism}.vcf.gz", dataset=dataset, chrom=chroms, allelism=['multiallelic', '.biallelic']),
        expand("results/relatedness/ngsRelate_{chrom}", chrom=chroms),

include: "rules/utility.smk"
include: "rules/qc.smk"
include: "rules/relatedness.smk"
include: "rules/selection.smk"


### PLANS


# PCA 
# Selection - G12 or H12 if phased, XP-EHH, PBS
# Relatedness - ngsRelate
# Structure - fineStructure, MAPS
# mutations - 