## Snakefile for genomic surviellance of Anopheles gambiae
import pandas as pd
import numpy as np

configfile: "config/config.yaml"
include: "rules/common.smk"

# dataset
contigs = config['contigs']
dataset = config['dataset']

# metadata
metadata = pd.read_csv(config['metadata'], sep="\t")
#print(metadata)
#print(metadata.columns)

# Determine cohorts
cohorts = getCohorts(metadata,  columns=config['metadataCohortColumns'], minPopSize=15)
PCAcohorts = getCohorts(metadata, columns=['species_gambiae_coluzzii'])

if config['Selection']['PBS']['activate']:
    PBScohorts = getCohorts(metadata, columns=config['metadataCohortColumns'], comparatorColumn=config['Selection']['PBS']['metadataComparatorColumn'], minPopSize=15)
    PBScohorts = PBScohorts.dropna()
else:
    PBScohorts = getCohorts(metadata,  columns=config['metadataCohortColumns'], minPopSize=15)


rule all:
    input:
#        expand("resources/vcfs/{dataset}_{contig}.{allelism}.vcf.gz", dataset=config['dataset'], contig=config['contigs'], allelism=['biallelic']),
#        expand("results/relatedness/ngsRelate.{dataset}.{contig}", contig=config['contigs'], dataset=config['dataset']) if config['PopulationStructure']['NgsRelate']['activate'] else [],
#        expand("results/selection/G123/G123_{cohort}.{contig}.png", cohort= cohorts['cohortNoSpaceText'], contig=contigs) if config['Selection']['G123']['activate'] else [],
        #expand("results/PCA/{cohort}.{contig}.html", cohort=PCAcohorts['cohortNoSpaceText'], contig=contigs),
       # expand("results/selection/PBS/PBS_{cohort}.{contig}.png", cohort=PBScohorts['cohortNoSpaceText'], contig=contigs), #if config['Selection']['PBS']['activate'] else [],
    #    expand("results/variantsOfInterest/VOI.{dataset}.frequencies.tsv", dataset = dataset),
    #    expand("results/variantsOfInterest/VOI.{dataset}.heatmap.png", dataset= dataset),
        expand("results/f2HapLengths_{contig}.tsv", contig=contigs)




include: "rules/utility.smk"
include: "rules/qc.smk"
include: "rules/PopulationStructure.smk"
include: "rules/selection.smk"
#include: "rules/karyotype.smk"

### plan for inputs 
# if Zarr present and VCF not, write lightweight VCF to file.
# if VCF present and Zarr not, write Zarrs to file
# Use Zarrs for all python analyses.
# Use VCFs for programs that require VCF
# How to determine name of VCF or Zarrs? use input function, if VCF provided use provided name, if 
# not provided, use default name. Same for Zarr. 
# provide metadata column name to split into populations (i.e location), optionally, provide multiple columns to split 
# into groups i.e location | species | year. 
# 
# Need to figure out what to with colours. Give colour column? 
#
# choose ngsRelate contigs 
# AgMONITOR
# 
# VOI - Bokeh plots, all VOIs, map of sample area showing predicted AFs
# some sort of 2d spatial interpolation of allele frequencies, or likelihood based geostatistic modelling
#
#
# PopulationStructure
# 
# - doubleton sharing, haplotype sharing, IBD tracts 
# 
# diversity 
# - conf intervals 
