{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b23f667a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import allel\n",
    "import phasedibd as ibd \n",
    "import malariagen_data\n",
    "import pandas as pd \n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93f4ddfa",
   "metadata": {},
   "source": [
    "### IBD detection with phasedIBD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e22706b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_sets = ['1244-VO-GH-YAWSON-VMF00149']\n",
    "analysis = 'gamb_colu'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "02324337",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/javascript": "(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  const force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\nconst JS_MIME_TYPE = 'application/javascript';\n  const HTML_MIME_TYPE = 'text/html';\n  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n  const CLASS_NAME = 'output_bokeh rendered_html';\n\n  /**\n   * Render data to the DOM node\n   */\n  function render(props, node) {\n    const script = document.createElement(\"script\");\n    node.appendChild(script);\n  }\n\n  /**\n   * Handle when an output is cleared or removed\n   */\n  function handleClearOutput(event, handle) {\n    const cell = handle.cell;\n\n    const id = cell.output_area._bokeh_element_id;\n    const server_id = cell.output_area._bokeh_server_id;\n    // Clean up Bokeh references\n    if (id != null && id in Bokeh.index) {\n      Bokeh.index[id].model.document.clear();\n      delete Bokeh.index[id];\n    }\n\n    if (server_id !== undefined) {\n      // Clean up Bokeh references\n      const cmd_clean = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n      cell.notebook.kernel.execute(cmd_clean, {\n        iopub: {\n          output: function(msg) {\n            const id = msg.content.text.trim();\n            if (id in Bokeh.index) {\n              Bokeh.index[id].model.document.clear();\n              delete Bokeh.index[id];\n            }\n          }\n        }\n      });\n      // Destroy server and session\n      const cmd_destroy = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n      cell.notebook.kernel.execute(cmd_destroy);\n    }\n  }\n\n  /**\n   * Handle when a new output is added\n   */\n  function handleAddOutput(event, handle) {\n    const output_area = handle.output_area;\n    const output = handle.output;\n\n    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n      return\n    }\n\n    const toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n\n    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n      // store reference to embed id on output_area\n      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n    }\n    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n      const bk_div = document.createElement(\"div\");\n      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n      const script_attrs = bk_div.children[0].attributes;\n      for (let i = 0; i < script_attrs.length; i++) {\n        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n      }\n      // store reference to server id on output_area\n      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n    }\n  }\n\n  function register_renderer(events, OutputArea) {\n\n    function append_mime(data, metadata, element) {\n      // create a DOM node to render to\n      const toinsert = this.create_output_subarea(\n        metadata,\n        CLASS_NAME,\n        EXEC_MIME_TYPE\n      );\n      this.keyboard_manager.register_events(toinsert);\n      // Render to node\n      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n      render(props, toinsert[toinsert.length - 1]);\n      element.append(toinsert);\n      return toinsert\n    }\n\n    /* Handle when an output is cleared or removed */\n    events.on('clear_output.CodeCell', handleClearOutput);\n    events.on('delete.Cell', handleClearOutput);\n\n    /* Handle when a new output is added */\n    events.on('output_added.OutputArea', handleAddOutput);\n\n    /**\n     * Register the mime type and append_mime function with output_area\n     */\n    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n      /* Is output safe? */\n      safe: true,\n      /* Index of renderer in `output_area.display_order` */\n      index: 0\n    });\n  }\n\n  // register the mime type if in Jupyter Notebook environment and previously unregistered\n  if (root.Jupyter !== undefined) {\n    const events = require('base/js/events');\n    const OutputArea = require('notebook/js/outputarea').OutputArea;\n\n    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n      register_renderer(events, OutputArea);\n    }\n  }\n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  const NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    const el = document.getElementById(null);\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-2.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-2.4.3.min.js\"];\n  const css_urls = [];\n\n  const inline_js = [    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\nfunction(Bokeh) {\n    }\n  ];\n\n  function run_inline_js() {\n    if (root.Bokeh !== undefined || force === true) {\n          for (let i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\n} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      const cell = $(document.getElementById(null)).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));",
      "application/vnd.bokehjs_load.v0+json": ""
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ag3 = malariagen_data.Ag3(pre=True, results_cache=\"../../malariagen_data_cache\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fb39bf68",
   "metadata": {},
   "outputs": [],
   "source": [
    "chromosomes = list(ag3.virtual_contigs) + ['X']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b7cf47e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def tpbwt_ibd(region, analysis, sample_sets, sample_query=None, L_m=300, L_f=3, missing_site_threshold=10, segments_out_path=None, verbose=True):\n",
    "    \n",
    "    print(\"loading haplotypes\")\n",
    "    region = ag3.resolve_region(region)\n",
    "    ds_haps = ag3.haplotypes(region=region, analysis=analysis, sample_sets=sample_sets, sample_query=sample_query)\n",
    "    gt = allel.GenotypeDaskArray(ds_haps[\"call_genotype\"].data)\n",
    "    \n",
    "    print(\"computing dask array\")\n",
    "    ht = gt.to_haplotypes().compute()\n",
    "    pos = ds_haps['variant_position'].values\n",
    "    \n",
    "    print(f\"loading haplotypes into tpbwt for {region}\")\n",
    "    haplotypes = ibd.HaplotypeAlignment(haplotype_array=ht.T, chromosomes=[region.contig])\n",
    "    tpbwt = ibd.TPBWTAnalysis()\n",
    "    \n",
    "    print(f\"computing ibd segments with templated positional burrows wheeler transform\")\n",
    "    ibd_df = tpbwt.compute_ibd(haplotypes, L_m=L_m, L_f=L_f, missing_site_threshold=missing_site_threshold, verbose=verbose, segments_out_path=segments_out_path)\n",
    "    ibd_df = ibd_df.assign(start_bp=lambda x: pos[x.start_bp], end_bp=lambda x: pos[x.end_bp])\n",
    "    ibd_df = ibd_df.assign(size=ibd_df.end_bp - ibd_df.start_bp)\n",
    "    print(\"complete\")\n",
    "\n",
    "    return ibd_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6e0b7644",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loading haplotypes\n",
      "computing dask array\n",
      "loading haplotypes into tpbwt for 2L\n",
      "computing ibd segments with templated positional burrows wheeler transform\n",
      "\n",
      "IBD compute for chromosome 2L...\n",
      "Running in-sample TPBWT on 14810165 sites.\n",
      "Number of haplotypes: 970\n",
      "Minimum memory required: 0.0075272 Gb\n",
      "Increasing size of segment vectors...\n",
      "Increasing size of segment vectors...\n",
      "Increasing size of segment vectors...\n",
      "Increasing size of segment vectors...\n",
      "\n",
      "Done computing IBD segments. Finishing up...\n",
      "Number of IBD segments found = 121873367\n",
      "Building final dataframe...\n"
     ]
    },
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mCanceled future for execute_request message before replies were done"
     ]
    },
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mThe Kernel crashed while executing code in the the current cell or a previous cell. Please review the code in the cell(s) to identify a possible cause of the failure. Click <a href='https://aka.ms/vscodeJupyterKernelCrash'>here</a> for more info. View Jupyter <a href='command:jupyter.viewOutput'>log</a> for further details."
     ]
    }
   ],
   "source": [
    "for contig in ag3.contigs:\n",
    "\n",
    "    ibd_df = tpbwt_ibd(\n",
    "        region=contig, \n",
    "        analysis=\"gamb_colu\", \n",
    "        sample_sets=sample_sets,\n",
    "        segments_out_path=f\"../../results/ibd_segments_{contig}.tsv\",\n",
    "        L_f=30\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "184cb69a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>id1</th>\n",
       "      <th>id2</th>\n",
       "      <th>id1_haplotype</th>\n",
       "      <th>id2_haplotype</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>start_cm</th>\n",
       "      <th>end_cm</th>\n",
       "      <th>start_bp</th>\n",
       "      <th>end_bp</th>\n",
       "      <th>size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2L</td>\n",
       "      <td>968</td>\n",
       "      <td>969</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7785</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7785.0</td>\n",
       "      <td>1249</td>\n",
       "      <td>179152</td>\n",
       "      <td>177903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2L</td>\n",
       "      <td>966</td>\n",
       "      <td>969</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7788</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7788.0</td>\n",
       "      <td>1249</td>\n",
       "      <td>179158</td>\n",
       "      <td>177909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2L</td>\n",
       "      <td>962</td>\n",
       "      <td>969</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7794</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7794.0</td>\n",
       "      <td>1249</td>\n",
       "      <td>179172</td>\n",
       "      <td>177923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2L</td>\n",
       "      <td>958</td>\n",
       "      <td>969</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7793</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7793.0</td>\n",
       "      <td>1249</td>\n",
       "      <td>179171</td>\n",
       "      <td>177922</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2L</td>\n",
       "      <td>954</td>\n",
       "      <td>969</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7793</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7793.0</td>\n",
       "      <td>1249</td>\n",
       "      <td>179171</td>\n",
       "      <td>177922</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>302794</th>\n",
       "      <td>2L</td>\n",
       "      <td>965</td>\n",
       "      <td>969</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7793</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7793.0</td>\n",
       "      <td>1249</td>\n",
       "      <td>179171</td>\n",
       "      <td>177922</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>302795</th>\n",
       "      <td>2L</td>\n",
       "      <td>966</td>\n",
       "      <td>967</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8715</td>\n",
       "      <td>38345</td>\n",
       "      <td>8715.0</td>\n",
       "      <td>38345.0</td>\n",
       "      <td>185468</td>\n",
       "      <td>496807</td>\n",
       "      <td>311339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>302796</th>\n",
       "      <td>2L</td>\n",
       "      <td>966</td>\n",
       "      <td>969</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8677</td>\n",
       "      <td>38537</td>\n",
       "      <td>8677.0</td>\n",
       "      <td>38537.0</td>\n",
       "      <td>185343</td>\n",
       "      <td>497476</td>\n",
       "      <td>312133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>302797</th>\n",
       "      <td>2L</td>\n",
       "      <td>967</td>\n",
       "      <td>968</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8677</td>\n",
       "      <td>38537</td>\n",
       "      <td>8677.0</td>\n",
       "      <td>38537.0</td>\n",
       "      <td>185343</td>\n",
       "      <td>497476</td>\n",
       "      <td>312133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>302798</th>\n",
       "      <td>2L</td>\n",
       "      <td>968</td>\n",
       "      <td>969</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8676</td>\n",
       "      <td>38343</td>\n",
       "      <td>8676.0</td>\n",
       "      <td>38343.0</td>\n",
       "      <td>185338</td>\n",
       "      <td>496801</td>\n",
       "      <td>311463</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>302799 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chromosome  id1  id2  id1_haplotype  id2_haplotype  start    end  \\\n",
       "0              2L  968  969              0              0      0   7785   \n",
       "1              2L  966  969              0              0      0   7788   \n",
       "2              2L  962  969              0              0      0   7794   \n",
       "3              2L  958  969              0              0      0   7793   \n",
       "4              2L  954  969              0              0      0   7793   \n",
       "...           ...  ...  ...            ...            ...    ...    ...   \n",
       "302794         2L  965  969              0              0      0   7793   \n",
       "302795         2L  966  967              0              0   8715  38345   \n",
       "302796         2L  966  969              0              0   8677  38537   \n",
       "302797         2L  967  968              0              0   8677  38537   \n",
       "302798         2L  968  969              0              0   8676  38343   \n",
       "\n",
       "        start_cm   end_cm  start_bp  end_bp    size  \n",
       "0            0.0   7785.0      1249  179152  177903  \n",
       "1            0.0   7788.0      1249  179158  177909  \n",
       "2            0.0   7794.0      1249  179172  177923  \n",
       "3            0.0   7793.0      1249  179171  177922  \n",
       "4            0.0   7793.0      1249  179171  177922  \n",
       "...          ...      ...       ...     ...     ...  \n",
       "302794       0.0   7793.0      1249  179171  177922  \n",
       "302795    8715.0  38345.0    185468  496807  311339  \n",
       "302796    8677.0  38537.0    185343  497476  312133  \n",
       "302797    8677.0  38537.0    185343  497476  312133  \n",
       "302798    8676.0  38343.0    185338  496801  311463  \n",
       "\n",
       "[302799 rows x 12 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ibd_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3628709",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loading haplotypes\n",
      "computing dask array\n",
      "loading haplotypes into tpbwt for 2R\n",
      "computing ibd segments with templated positional burrows wheeler transform\n",
      "\n",
      "IBD compute for chromosome 2R...\n"
     ]
    }
   ],
   "source": [
    "ibd_dfs = []\n",
    "\n",
    "for chrom in ag3.contigs:\n",
    "    \n",
    "    ibd_df = tpbwt_ibd(\n",
    "        contig=chrom, \n",
    "        analysis=analysis, \n",
    "        sample_sets=sample_sets,\n",
    "        L_f=50_000\n",
    "    )\n",
    "    print(\"complete\")\n",
    "    ibd_dfs.append(ibd_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3af74d92",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
