{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "#import malariagen_data\n",
    "import allel\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import dask\n",
    "from dask.diagnostics import ProgressBar\n",
    "import zarr\n",
    "from pathlib import Path\n",
    "import scipy\n",
    "import seaborn as sns\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "%run tools.py"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Karyotyping the GAARDIAN data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "ag3 = malariagen_data.Ag3(\"gs://vo_agam_release/\", pre=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_set</th>\n",
       "      <th>sample_count</th>\n",
       "      <th>release</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1191-VO-MULTI-OLOUGHLIN-VMF00106</td>\n",
       "      <td>237</td>\n",
       "      <td>v3.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1191-VO-MULTI-OLOUGHLIN-VMF00140</td>\n",
       "      <td>1095</td>\n",
       "      <td>v3.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1244-VO-GH-YAWSON-VMF00149</td>\n",
       "      <td>485</td>\n",
       "      <td>v3.4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         sample_set  sample_count release\n",
       "0  1191-VO-MULTI-OLOUGHLIN-VMF00106           237    v3.4\n",
       "1  1191-VO-MULTI-OLOUGHLIN-VMF00140          1095    v3.4\n",
       "2        1244-VO-GH-YAWSON-VMF00149           485    v3.4"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sample_sets = ag3.sample_sets(release=\"v3.4\")\n",
    "df_sample_sets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "my_sample_set = '1244-VO-GH-YAWSON-VMF00149'\n",
    "metadata = ag3.sample_metadata(sample_sets=my_sample_set)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata['location2'] = metadata['location'].str.split(\".\").str.get(0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## compkaryo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- Running CompKaryo 2La--- \n",
      "--- Running CompKaryo 2Rj--- \n",
      "--- Running CompKaryo 2Rb--- \n",
      "--- Running CompKaryo 2Rc_col--- \n",
      "--- Running CompKaryo 2Rc_gam--- \n",
      "--- Running CompKaryo 2Rd--- \n",
      "--- Running CompKaryo 2Ru--- \n"
     ]
    }
   ],
   "source": [
    "invDict = {}\n",
    "\n",
    "for inversion in inversionDict.keys():\n",
    "    chrom = inversionDict[inversion][0]\n",
    "\n",
    "    snps = allel.GenotypeDaskArray(ag3.snp_genotypes(contig=chrom, sample_sets=my_sample_set))\n",
    "    pos = ag3.snp_sites(contig=chrom)\n",
    "    pos = allel.SortedIndex(pos[0])\n",
    "\n",
    "    callset = {'geno':snps, 'pos':pos, 'chrom':chrom}\n",
    "    print(f\"--- Running CompKaryo {inversion}--- \")\n",
    "    av_gts, total_sites, num_0, num_1, num_2 = compkaryo(callset, inversion)\n",
    "    total_sites = total_sites.compute()\n",
    "    invDict[inversion] = pd.DataFrame({'partner_sample_id': metadata['sample_id'], \n",
    "                           'inversion':inversion, \n",
    "                           'mean_genotype': av_gts, \n",
    "                           'total_snp_tags':total_sites})\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = metadata[['sample_id', 'partner_sample_id', 'location2', 'species_gambiae_coluzzii']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gaardian_karyotypes = pd.concat(invDict.values()).rename(columns={'partner_sample_id':'sample_id'})\n",
    "gaardian_karyotypes = gaardian_karyotypes.merge(df).set_index('partner_sample_id').reset_index().drop(columns='sample_id')\n",
    "gaardian_karyotypes.to_csv(\"gaardian_karyotypes.tsv\", sep=\"\\t\") "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load karyo data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "gaardian_karyotypes = pd.read_csv(\"../gaardian_karyotypes.tsv\", sep=\"\\t\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>partner_sample_id</th>\n",
       "      <th>inversion</th>\n",
       "      <th>mean_genotype</th>\n",
       "      <th>total_snp_tags</th>\n",
       "      <th>location2</th>\n",
       "      <th>species_gambiae_coluzzii</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WA-2001</td>\n",
       "      <td>2La</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>209</td>\n",
       "      <td>Adumanu</td>\n",
       "      <td>gambiae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WA-2001</td>\n",
       "      <td>2Rj</td>\n",
       "      <td>0.060606</td>\n",
       "      <td>99</td>\n",
       "      <td>Adumanu</td>\n",
       "      <td>gambiae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WA-2001</td>\n",
       "      <td>2Rb</td>\n",
       "      <td>0.123209</td>\n",
       "      <td>349</td>\n",
       "      <td>Adumanu</td>\n",
       "      <td>gambiae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WA-2001</td>\n",
       "      <td>2Rc_col</td>\n",
       "      <td>0.438596</td>\n",
       "      <td>57</td>\n",
       "      <td>Adumanu</td>\n",
       "      <td>gambiae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WA-2001</td>\n",
       "      <td>2Rc_gam</td>\n",
       "      <td>0.061224</td>\n",
       "      <td>49</td>\n",
       "      <td>Adumanu</td>\n",
       "      <td>gambiae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3390</th>\n",
       "      <td>WA-2500</td>\n",
       "      <td>2Rb</td>\n",
       "      <td>0.086207</td>\n",
       "      <td>348</td>\n",
       "      <td>Domenase</td>\n",
       "      <td>coluzzii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3391</th>\n",
       "      <td>WA-2500</td>\n",
       "      <td>2Rc_col</td>\n",
       "      <td>0.122807</td>\n",
       "      <td>57</td>\n",
       "      <td>Domenase</td>\n",
       "      <td>coluzzii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3392</th>\n",
       "      <td>WA-2500</td>\n",
       "      <td>2Rc_gam</td>\n",
       "      <td>0.020408</td>\n",
       "      <td>49</td>\n",
       "      <td>Domenase</td>\n",
       "      <td>coluzzii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3393</th>\n",
       "      <td>WA-2500</td>\n",
       "      <td>2Rd</td>\n",
       "      <td>0.081633</td>\n",
       "      <td>147</td>\n",
       "      <td>Domenase</td>\n",
       "      <td>coluzzii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3394</th>\n",
       "      <td>WA-2500</td>\n",
       "      <td>2Ru</td>\n",
       "      <td>0.033898</td>\n",
       "      <td>177</td>\n",
       "      <td>Domenase</td>\n",
       "      <td>coluzzii</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3395 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     partner_sample_id inversion  mean_genotype  total_snp_tags location2  \\\n",
       "0              WA-2001       2La       0.000000             209   Adumanu   \n",
       "1              WA-2001       2Rj       0.060606              99   Adumanu   \n",
       "2              WA-2001       2Rb       0.123209             349   Adumanu   \n",
       "3              WA-2001   2Rc_col       0.438596              57   Adumanu   \n",
       "4              WA-2001   2Rc_gam       0.061224              49   Adumanu   \n",
       "...                ...       ...            ...             ...       ...   \n",
       "3390           WA-2500       2Rb       0.086207             348  Domenase   \n",
       "3391           WA-2500   2Rc_col       0.122807              57  Domenase   \n",
       "3392           WA-2500   2Rc_gam       0.020408              49  Domenase   \n",
       "3393           WA-2500       2Rd       0.081633             147  Domenase   \n",
       "3394           WA-2500       2Ru       0.033898             177  Domenase   \n",
       "\n",
       "     species_gambiae_coluzzii  \n",
       "0                     gambiae  \n",
       "1                     gambiae  \n",
       "2                     gambiae  \n",
       "3                     gambiae  \n",
       "4                     gambiae  \n",
       "...                       ...  \n",
       "3390                 coluzzii  \n",
       "3391                 coluzzii  \n",
       "3392                 coluzzii  \n",
       "3393                 coluzzii  \n",
       "3394                 coluzzii  \n",
       "\n",
       "[3395 rows x 6 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gaardian_karyotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_karyotype_freq(dfpath, inversion, species, metadata):\n",
    "    \n",
    "    karyotypes = pd.read_csv(dfpath, sep=\"\\t\", index_col=0)\n",
    "    karyo = karyotypes.loc[(karyotypes['species_gambiae_coluzzii'] == species) & (karyotypes['inversion'] == inversion)]\n",
    "    karyo['karyotype']  = karyo.mean_genotype.round()\n",
    "    \n",
    "    df = pd.DataFrame(karyo.groupby(['inversion', 'location2']).karyotype.value_counts()).rename(columns={'karyotype':'n'}).reset_index()\n",
    "    df['total_alleles'] = df['karyotype']*df['n']\n",
    "    df = df.groupby('location2').agg({'n':'sum', 'total_alleles':'sum'})\n",
    "    df[f'{inversion}_frequency'] = df['total_alleles']/(df['n']*2)\n",
    "    df = df.drop(columns=['total_alleles']).reset_index()\n",
    "\n",
    "    metalocs = metadata[['location2', 'latitude', 'longitude']].groupby('location2').agg({'latitude':'mean', 'longitude':'mean'}).reset_index()\n",
    "\n",
    "    \n",
    "    karyo_meta = pd.merge(df, metalocs)\n",
    "    return  karyo_meta\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sanj/apps/anaconda3/envs/pysanj/lib/python3.7/site-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \"\"\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>location2</th>\n",
       "      <th>n</th>\n",
       "      <th>2Rb_frequency</th>\n",
       "      <th>latitude</th>\n",
       "      <th>longitude</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Adansi Apagya</td>\n",
       "      <td>15</td>\n",
       "      <td>0.033333</td>\n",
       "      <td>5.991625</td>\n",
       "      <td>-1.371125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Adansi-Krom</td>\n",
       "      <td>6</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.170833</td>\n",
       "      <td>-1.835500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Adumanu</td>\n",
       "      <td>5</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.303333</td>\n",
       "      <td>-1.681500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Anhwiaso</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.144000</td>\n",
       "      <td>-1.469000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Ankaako</td>\n",
       "      <td>12</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>6.175833</td>\n",
       "      <td>-1.793083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Annorkrom</td>\n",
       "      <td>46</td>\n",
       "      <td>0.021739</td>\n",
       "      <td>5.970000</td>\n",
       "      <td>-1.697000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Anwona</td>\n",
       "      <td>5</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>6.089000</td>\n",
       "      <td>-1.516250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Bogyawe</td>\n",
       "      <td>3</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.388800</td>\n",
       "      <td>-1.548000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Domenase</td>\n",
       "      <td>57</td>\n",
       "      <td>0.017544</td>\n",
       "      <td>6.047947</td>\n",
       "      <td>-1.961877</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Dompoase</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.308667</td>\n",
       "      <td>-1.533333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Fumso</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.107500</td>\n",
       "      <td>-1.449250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Jacobu</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.345000</td>\n",
       "      <td>-1.672000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Kente</td>\n",
       "      <td>23</td>\n",
       "      <td>0.021739</td>\n",
       "      <td>6.299500</td>\n",
       "      <td>-1.801208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Kokotro</td>\n",
       "      <td>24</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.369393</td>\n",
       "      <td>-1.462964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Koniyaw</td>\n",
       "      <td>31</td>\n",
       "      <td>0.016129</td>\n",
       "      <td>6.404094</td>\n",
       "      <td>-1.452000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Mprakyire</td>\n",
       "      <td>9</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.949800</td>\n",
       "      <td>-1.548800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>New Edubiase</td>\n",
       "      <td>15</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.058881</td>\n",
       "      <td>-1.391048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Nkotumso</td>\n",
       "      <td>30</td>\n",
       "      <td>0.016667</td>\n",
       "      <td>6.000267</td>\n",
       "      <td>-1.919767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Odumto</td>\n",
       "      <td>32</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>6.254324</td>\n",
       "      <td>-1.841000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Pokukrom</td>\n",
       "      <td>37</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.015676</td>\n",
       "      <td>-1.812486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Prisintease</td>\n",
       "      <td>24</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.970667</td>\n",
       "      <td>-1.778500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Subin Camp</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.966500</td>\n",
       "      <td>-1.451500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Wamase</td>\n",
       "      <td>16</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.085062</td>\n",
       "      <td>-1.697000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Watreso</td>\n",
       "      <td>19</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.253316</td>\n",
       "      <td>-1.916947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Yadome</td>\n",
       "      <td>6</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.048500</td>\n",
       "      <td>-1.662833</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        location2   n  2Rb_frequency  latitude  longitude\n",
       "0   Adansi Apagya  15       0.033333  5.991625  -1.371125\n",
       "1     Adansi-Krom   6       0.000000  6.170833  -1.835500\n",
       "2         Adumanu   5       0.000000  6.303333  -1.681500\n",
       "3        Anhwiaso   2       0.000000  6.144000  -1.469000\n",
       "4         Ankaako  12       0.083333  6.175833  -1.793083\n",
       "5       Annorkrom  46       0.021739  5.970000  -1.697000\n",
       "6          Anwona   5       0.100000  6.089000  -1.516250\n",
       "7         Bogyawe   3       0.000000  6.388800  -1.548000\n",
       "8        Domenase  57       0.017544  6.047947  -1.961877\n",
       "9        Dompoase   1       0.000000  6.308667  -1.533333\n",
       "10          Fumso   1       0.000000  6.107500  -1.449250\n",
       "11         Jacobu   2       0.000000  6.345000  -1.672000\n",
       "12          Kente  23       0.021739  6.299500  -1.801208\n",
       "13        Kokotro  24       0.000000  6.369393  -1.462964\n",
       "14        Koniyaw  31       0.016129  6.404094  -1.452000\n",
       "15      Mprakyire   9       0.000000  5.949800  -1.548800\n",
       "16   New Edubiase  15       0.000000  6.058881  -1.391048\n",
       "17       Nkotumso  30       0.016667  6.000267  -1.919767\n",
       "18         Odumto  32       0.031250  6.254324  -1.841000\n",
       "19       Pokukrom  37       0.000000  6.015676  -1.812486\n",
       "20    Prisintease  24       0.000000  5.970667  -1.778500\n",
       "21     Subin Camp   1       0.000000  5.966500  -1.451500\n",
       "22         Wamase  16       0.000000  6.085062  -1.697000\n",
       "23        Watreso  19       0.000000  6.253316  -1.916947\n",
       "24         Yadome   6       0.000000  6.048500  -1.662833"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_karyotype_freq(\"../gaardian_karyotypes.tsv\", \"2Rb\", \"coluzzii\", metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.7.3 64-bit ('pysanj': conda)",
   "language": "python",
   "name": "python37364bitpysanjconda39958c5ba9ef41c3a5542f2bfe81dc48"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
