{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46e1af42",
   "metadata": {
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "sample_set = \"\"\n",
    "release = \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53ad9ee1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import google.auth\n",
    "import gcsfs\n",
    "\n",
    "credentials, project_id = google.auth.default()\n",
    "print(credentials)\n",
    "print(project_id)\n",
    "\n",
    "gcs = gcsfs.GCSFileSystem(\n",
    "      token=credentials,\n",
    "      project=project_id, \n",
    "      cache_timeout=0, \n",
    "      block_size=2**18,\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0aec5ef-e85f-48e4-8478-6f8c62d3cf0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import ancIBD\n",
    "import allel\n",
    "import malariagen_data\n",
    "import dask \n",
    "\n",
    "dask.config.set(num_workers=4)\n",
    "\n",
    "import pandas as pd\n",
    "import itertools\n",
    "from tqdm.notebook import tqdm\n",
    "from numba import njit\n",
    "\n",
    "@njit()\n",
    "def convert_genotypes_to_probs(genotypes, error_rate=0.01):\n",
    "    snps, samples, _ = genotypes.shape\n",
    "    probs = np.zeros((snps, samples, 3), dtype=np.float64)\n",
    "    \n",
    "    for i in range(snps):\n",
    "        for j in range(samples):\n",
    "            allele1, allele2 = genotypes[i, j]\n",
    "            \n",
    "            if allele1 == 0 and allele2 == 0:\n",
    "                probs[i, j] = [1 - error_rate, error_rate / 2, error_rate / 2]\n",
    "            elif allele1 == 0 and allele2 == 1:\n",
    "                probs[i, j] = [error_rate / 2, 1 - error_rate, error_rate / 2]\n",
    "            elif allele1 == 1 and allele2 == 0:\n",
    "                probs[i, j] = [error_rate / 2, 1 - error_rate, error_rate / 2]\n",
    "            else:  # allele1 == 1 and allele2 == 1\n",
    "                probs[i, j] = [error_rate / 2, error_rate / 2, 1 - error_rate]\n",
    "    \n",
    "    return probs\n",
    "\n",
    "def read_ag_gmap(contig):\n",
    "    if contig in {\"2RL\", \"3RL\"}:\n",
    "        chrom = contig[0]\n",
    "        contig_r, contig_l = f\"{chrom}R\", f\"{chrom}L\"\n",
    "        df_r = read_ag_gmap(contig_r)\n",
    "        df_l = read_ag_gmap(contig_l)\n",
    "        max_ppos = df_r[\"pposition\"].iloc[-1]\n",
    "        max_gpos = df_r[\"gposition\"].iloc[-1]\n",
    "        df_l = df_l.iloc[1:]\n",
    "        df_l[\"pposition\"] += max_ppos\n",
    "        df_l[\"gposition\"] += max_gpos\n",
    "        df = pd.concat([df_r, df_l], axis=0, ignore_index=True)\n",
    "    else:\n",
    "        df = pd.read_csv(gmap_dict[contig], sep=\"\\t\")\n",
    "    return df\n",
    "    \n",
    "def ag_gmap(contig):\n",
    "    ag3 = malariagen_data.Ag3()\n",
    "        \n",
    "    # read in the genetic map file\n",
    "    df_gmap = read_ag_gmap(contig)\n",
    "\n",
    "    # set up an array of per-base recombination rate values\n",
    "    rr = np.zeros(len(ag3.genome_sequence(contig)), dtype=\"f8\")\n",
    "\n",
    "    # fill in the recombination rate values from the genetic map file\n",
    "    for row, next_row in zip(\n",
    "        itertools.islice(df_gmap.itertuples(), 0, len(df_gmap)-1), \n",
    "        itertools.islice(df_gmap.itertuples(), 1, None)):\n",
    "        \n",
    "        # N.B., the genetic map file is in units of cM / Mbp\n",
    "        # we multiple by 1e-6 to convert to cM / bp\n",
    "        rr[row.pposition-1:next_row.pposition] = row.rrate * 1e-6\n",
    "    \n",
    "    # compute mapping from physical to genetic position\n",
    "    gmap = np.cumsum(rr)\n",
    "        \n",
    "    return gmap\n",
    "    \n",
    "def ag_p2g(contig, ppos):\n",
    "    \"\"\"Convert physical position (bp) to genetic position (M).\"\"\"\n",
    "    gmap = ag_gmap(contig)\n",
    "    gpos = gmap[ppos - 1]\n",
    "    return gpos / 100 # morgans not centimorgans \n",
    "\n",
    "\n",
    "gmap_dict = {contig : f\"https://raw.githubusercontent.com/anopheles-genomic-surveillance/selection-atlas/main/workflow/notebooks/ag_{contig}.gmap\"\n",
    "            for contig in ['2L', '2R', '3L', '3R', 'X']\n",
    "}\n",
    "\n",
    "\n",
    "def prepare_ancIBD_input(contig, sample_sets, sample_query, cohort_size=None):\n",
    "#     import zarr\n",
    "#     gt = zarr.load(f\"results/crosses-gn-{contig}.zarr\")\n",
    "#     pos = zarr.load(f\"results/pos-{contig}.zarr\")\n",
    "#     samples = zarr.load(\"results/crosses-samples.zarr\")\n",
    "\n",
    "    # load haplotypes\n",
    "    ds_haps = ag3.haplotypes(region=contig, sample_sets=sample_sets, sample_query=sample_query, cohort_size=cohort_size)\n",
    "    gt = allel.GenotypeArray(ds_haps['call_genotype'].values)   \n",
    "    pos = ds_haps['variant_position'].values\n",
    "    samples = ds_haps['sample_id'].values\n",
    "\n",
    "    # find segregating sites\n",
    "    seg_ = gt.count_alleles().is_segregating()\n",
    "    gt = gt.compress(seg_, axis=0)\n",
    "    pos = pos[seg_]\n",
    "\n",
    "    # find the alternate allele frequency\n",
    "    frqs = gt.count_alleles().to_frequencies()\n",
    "    alt_frqs = frqs[:, 1]\n",
    "    \n",
    "    # convert GTs to genotype probabilities\n",
    "    gp = convert_genotypes_to_probs(gt.values)\n",
    "\n",
    "    # prepare input dictionary\n",
    "    input_dict = {\n",
    "        'calldata/GT':gt.values,\n",
    "        'calldata/GP':gp,\n",
    "        'calldata/AF': alt_frqs,\n",
    "        'variants/POS':pos,\n",
    "        'variants/MAP':ag_p2g(contig=contig, ppos=pos),\n",
    "        'samples':samples\n",
    "    }\n",
    "    \n",
    "    return input_dict\n",
    "\n",
    "# vectorized haversine function\n",
    "def haversine(lat1, lon1, lat2, lon2, to_radians=True, earth_radius=6371):\n",
    "    \"\"\"\n",
    "    slightly modified version: of http://stackoverflow.com/a/29546836/2901002\n",
    "\n",
    "    Calculate the great circle distance between two points\n",
    "    on the earth (specified in decimal degrees or in radians)\n",
    "\n",
    "    All (lat, lon) coordinates must have numeric dtypes and be of equal length.\n",
    "\n",
    "    \"\"\"\n",
    "    if to_radians:\n",
    "        lat1, lon1, lat2, lon2 = np.radians([lat1, lon1, lat2, lon2])\n",
    "\n",
    "    a = np.sin((lat2-lat1)/2.0)**2 + \\\n",
    "        np.cos(lat1) * np.cos(lat2) * np.sin((lon2-lon1)/2.0)**2\n",
    "\n",
    "    return earth_radius * 2 * np.arcsin(np.sqrt(a))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4336fb40-9398-4e69-978c-fad1440ae1c9",
   "metadata": {},
   "source": [
    "### I need to prepare a dict containing multiple arrays for a given contig:\n",
    "\n",
    "- calldata/GT\n",
    "- calldata/GP\n",
    "- genomic positions\n",
    "- linkage map\n",
    "- samples\n",
    "\n",
    "Lets try and create these directly from malariagen_data. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b41f195d-48c2-4ec2-ba9d-7f270f6b23ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "ag3 = malariagen_data.Ag3(results_cache=\"results_cache/\")\n",
    "\n",
    "# sample_sets = [\"AG1000G-GH\", \"AG1000G-BF-C\", \"AG1000G-TZ\"]\n",
    "sample_query = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df75abcb-4cad-4c83-bf8f-8ef93b5a1971",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "from ancIBD.run import hapBLOCK_chroms\n",
    "\n",
    "for contig in ('2RL', '3RL'):\n",
    "    print(\"Preparing input dict\")\n",
    "    input_dict = prepare_ancIBD_input(contig=contig, sample_sets=sample_set, sample_query=sample_query, cohort_size=None)\n",
    "    print(\"running ancIBD\")\n",
    "    df_ibd = hapBLOCK_chroms(folder_in=input_dict,\n",
    "                             iids=input_dict['samples'], \n",
    "                             run_iids=[],\n",
    "                             ch=contig, \n",
    "                             folder_out=f\"{release}_results/ibd/{sample_set}\",\n",
    "                             output=False, \n",
    "                             prefix_out='', \n",
    "                             logfile=False,\n",
    "                             l_model='dict',\n",
    "                             e_model='haploid_gl2', \n",
    "                             h_model='FiveStateScaled', \n",
    "                             t_model='standard',\n",
    "                             p_col='calldata/AF',\n",
    "                             ibd_in=1,\n",
    "                             ibd_out=10, \n",
    "                             ibd_jump=400,\n",
    "                             min_cm=8,\n",
    "                             cutoff_post=0.99,\n",
    "                             max_gap=0.0075)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0137ae90",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_ibd = df_ibd.merge(df_samples, left_on='iid1', right_on='sample_id').merge(df_samples, left_on='iid2', right_on='sample_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82ff0817",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_ibd.groupby(['iid1', 'iid2', 'taxon_x', 'taxon_y']).agg({'size':'sum'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "122896ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# from ancIBD.IO.ind_ibd import combine_all_chroms\n",
    "\n",
    "# combine_all_chroms(chs=('2RL', '3RL'),\n",
    "#                    folder_base='../results/species-ibd-',\n",
    "#                    path_save='../results/species_ibd.tsv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c730746",
   "metadata": {},
   "outputs": [],
   "source": [
    "# px.histogram(df_ibd_agg, x='size', color='taxon', template='simple_white')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
