{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75d449b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import zarr\n",
    "import numpy as np\n",
    "import plotly.express as px"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84bdcd65",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_zero_ibd_pairs(dfs_ibd, df_samples, fillna=0):\n",
    "\n",
    "    from itertools import combinations\n",
    "\n",
    "    dfs = [list(combinations(df_samples.query(\"sample_set == @s\")['sample_id'], 2)) for s in sample_sets]\n",
    "\n",
    "    comps = pd.concat([pd.DataFrame(d) for d in dfs], axis=0)\n",
    "    comps.columns = ['iid1', 'iid2']\n",
    "\n",
    "    def normalize_pairs(df, col1, col2):\n",
    "        df['normalized_iid1'] = df[[col1, col2]].min(axis=1)\n",
    "        df['normalized_iid2'] = df[[col1, col2]].max(axis=1)\n",
    "        return df\n",
    "\n",
    "    # Apply normalization\n",
    "    df_ibd_summary = normalize_pairs(dfs_ibd, 'iid1', 'iid2')\n",
    "    comps = normalize_pairs(comps, 'iid1', 'iid2')\n",
    "\n",
    "    df_ibd_summary = comps.merge(df_ibd_summary, how='left')\n",
    "\n",
    "    # Fill missing values with 0\n",
    "    df_ibd_summary.fillna({\n",
    "        'max_IBD': fillna,\n",
    "        'sum_IBD>8': 0,\n",
    "        'n_IBD>8': 0,\n",
    "        'sum_IBD>12': 0,\n",
    "        'n_IBD>12': 0,\n",
    "        'sum_IBD>16': 0,\n",
    "        'n_IBD>16': 0,\n",
    "        'sum_IBD>20': 0,\n",
    "    }, inplace=True)\n",
    "\n",
    "\n",
    "    df_ibd_summary = df_ibd_summary.merge(df_samples, left_on='iid1', right_on='sample_id').merge(df_samples, left_on='iid2', right_on='sample_id')\n",
    "    return df_ibd_summary"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd80b544",
   "metadata": {},
   "source": [
    "### Exploring IBD segments in ag3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3822f0d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c1eba00",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_sets1 = pd.Series(glob.glob(\"../../ag3_results/ibd/*/ch2RL.tsv\")).str.split(\"/\").str.get(-2).to_list()\n",
    "sample_sets2 = pd.Series(glob.glob(\"../../ag3_results/ibd/*/ch3RL.tsv\")).str.split(\"/\").str.get(-2).to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "714ed409",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_sets = [item for item in sample_sets1 if item in sample_sets2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2901efd",
   "metadata": {},
   "outputs": [],
   "source": [
    "from ancIBD.IO.ind_ibd import combine_all_chroms\n",
    "\n",
    "for sample_set in sample_sets:\n",
    "    combine_all_chroms(chs=('2RL', '3RL'),\n",
    "                       folder_base=f'../../ag3_results/ibd/{sample_set}/ch',\n",
    "                       path_save=f'../../ag3_results/ibd/{sample_set}/ibd.tsv'\n",
    "                      )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "945b79b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "from ancIBD.IO.ind_ibd import create_ind_ibd_df\n",
    "\n",
    "dfs = []\n",
    "for sample_set in sample_sets:\n",
    "    dfs.append(create_ind_ibd_df(ibd_data = f'../../ag3_results/ibd/{sample_set}/ibd.tsv',\n",
    "                          min_cms = [8, 12, 16, 20], snp_cm = 1000, min_cm = 8, sort_col = 0,\n",
    "                          savepath = f'../../ag3_results/ibd/{sample_set}/ibd_d220.tsv').assign(sample_set=sample_set\n",
    "                                                                                            ))\n",
    "    \n",
    "dfs_ibd = pd.concat(dfs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4eac2d5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import malariagen_data\n",
    "import matplotlib.pyplot as plt\n",
    "ag3 = malariagen_data.Ag3()\n",
    "df_samples = ag3.sample_metadata(sample_sets)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81c4cc93",
   "metadata": {},
   "source": [
    "Check that we now have the correct number of pairwise comparisons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e41a35b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_samples = df_samples.assign(set_taxon=lambda x: x.sample_set + \"-\" + x.taxon)\n",
    "\n",
    "from itertools import combinations\n",
    "\n",
    "# add zero ibd pairs\n",
    "df_ibd_summary = add_zero_ibd_pairs(dfs_ibd, df_samples, fillna=0).rename(columns={'max_IBD':'total_ibd_cm'})\n",
    "# assign a taxon + sample set column\n",
    "df_ibd_summary = df_ibd_summary.assign(set_taxon=[y if y == df_ibd_summary.set_taxon_x[i] else 'between-species' for i,y in enumerate(df_ibd_summary.set_taxon_y)])\n",
    "# remove combos with few observations\n",
    "df_ibd_summary = df_ibd_summary.groupby('set_taxon').filter(lambda g: len(g) >= 100)\n",
    "\n",
    "assert df_ibd_summary.query(\"sample_set == 'AG1000G-GH'\").shape[0] == len(list(combinations(df_samples.query(\"sample_set == 'AG1000G-GH'\").sample_id, 2)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d39d745",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"plotting\")\n",
    "fig = px.box(\n",
    "    df_ibd_summary.sort_values('set_taxon'),\n",
    "    y='total_ibd_cm', \n",
    "    x='set_taxon',\n",
    "    template='simple_white', \n",
    "    color='sample_set', \n",
    "    title='total shared IBD within sample sets / taxon combo',\n",
    "    height=700, \n",
    "    width=1000\n",
    ")\n",
    "fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b753670",
   "metadata": {},
   "outputs": [],
   "source": [
    "ibd_rate = df_ibd_summary.groupby('set_taxon').agg({'total_ibd_cm':'sum'}).assign(ibd_rate=lambda x: x.total_ibd_cm / df_ibd_summary.groupby('set_taxon').size())\n",
    "ibd_rate = ibd_rate.reset_index()\n",
    "\n",
    "fig = px.bar(\n",
    "    ibd_rate.sort_values('set_taxon'),\n",
    "    y='ibd_rate', \n",
    "    x='set_taxon',\n",
    "    template='simple_white', \n",
    "    title='Distribution of total shared IBD between individuals in ag3',\n",
    "    height=500, \n",
    "    width=800\n",
    ")\n",
    "fig"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6563a88a",
   "metadata": {},
   "source": [
    "### Plottijng with bokeh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92364123",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_ibd_segments_contig(\n",
    "    self,\n",
    "    df,\n",
    "    contig, \n",
    "    filter_cm=None,\n",
    "    show=False,\n",
    "    width=None,\n",
    "    height=400\n",
    "):\n",
    "    from itertools import combinations\n",
    "    import bokeh.plotting as bkplt\n",
    "    import bokeh.models as bkmod\n",
    "    import bokeh.layouts as bklay\n",
    "    \n",
    "    df = df.query(\"ch == @contig\")\n",
    "    \n",
    "    if filter_cm:\n",
    "        df = df.query(\"cm > @filter_cm\")\n",
    "        \n",
    "    print(\"finding levels\")\n",
    "    level_df = df[['iid1', 'iid2']].drop_duplicates()\n",
    "    # Add the 'level' column with values from 0 to 1\n",
    "    level_df['level'] = np.linspace(0, 1, level_df.shape[0])\n",
    "    # # Join with the original dataframe\n",
    "    df = df.merge(level_df, on=['iid1', 'iid2'], how='left')\n",
    "    \n",
    "#     print(\"colour mapping\")\n",
    "#     colour_mapping = {'half-ibd':'gray', \n",
    "#                       'full-ibd':'blue'}\n",
    "#     colour = df['ibd_type'].apply(lambda x: colour_mapping[x])\n",
    "\n",
    "    source = bkmod.ColumnDataSource(data={\n",
    "         'sample_id1':df['iid1'].to_numpy(),\n",
    "         'sample_id2':df['iid2'].to_numpy(),\n",
    "        'chromosome': df['ch'].to_numpy(),\n",
    "        'start': df['StartBP'].to_numpy(),\n",
    "        'end': df['EndBP'].to_numpy(),\n",
    "        'cm_length': df['lengthM'].to_numpy(),\n",
    "        'bottoms':df['level'].to_numpy(),\n",
    "        'tops': df['level'].to_numpy()+0.0001,\n",
    "#         'colour': colour.to_numpy()\n",
    "    })\n",
    "\n",
    "    hover = bkmod.HoverTool(tooltips=[\n",
    "             (\"sample_id1\", '@sample_id1'),\n",
    "             (\"sample_id2\", '@sample_id2'),\n",
    "             ('cm_length', '@cm_length'),\n",
    "            (\"segment span\", \"@start{,} - @end{,}\"),\n",
    "        ])\n",
    "        \n",
    "    print(\"making figure\")\n",
    "    if not width:\n",
    "        width = int(self.genome_sequence(contig).shape[0]/200000)\n",
    "    fig1 = bkplt.figure(title=contig,\n",
    "                        width=width,\n",
    "                        height=500, \n",
    "                        tools=\"tap,box_zoom,xpan,xzoom_in,xzoom_out,xwheel_zoom,reset\".split() + [hover],\n",
    "                        toolbar_location='above', active_drag='xpan', active_scroll='xwheel_zoom',\n",
    "                        output_backend=\"webgl\")\n",
    "\n",
    "    glyph = bkmod.Quad(left='start', right='end', bottom='bottoms', top='tops', line_color=\"grey\", line_alpha=.8, line_width=1)\n",
    "    fig1.add_glyph(source, glyph)\n",
    "\n",
    "    fig1.x_range = bkmod.Range1d(0, self.genome_sequence(contig).shape[0], bounds='auto')\n",
    "    fig1.y_range = bkmod.Range1d(0, 1, bounds='auto')\n",
    "    fig1.x_range.max_interval = self.genome_sequence(contig).shape[0]\n",
    "    fig1.yaxis.visible = False\n",
    "    fig1.ygrid.visible = False\n",
    "    _bokeh_style_genome_xaxis(fig1, contig)\n",
    "    \n",
    "    if show:\n",
    "        bkplt.show(fig1)\n",
    "    \n",
    "    return fig1\n",
    "\n",
    "def plot_ibd_segments(\n",
    "        self,\n",
    "        df, \n",
    "        out_dir,\n",
    "        cohort_id,\n",
    "        contigs=('2RL', '3RL', 'X'),\n",
    "        filter_cm=None,\n",
    "        show=True,\n",
    "        title=None,\n",
    "        height=1000\n",
    "    ):\n",
    "    import bokeh.models as bkmod\n",
    "    import bokeh.layouts as bklay\n",
    "    import bokeh.plotting as bkplt\n",
    "    from tqdm.notebook import tqdm\n",
    "    \n",
    "    figs = [\n",
    "            plot_ibd_segments_contig(\n",
    "                self=self,\n",
    "                df=df,\n",
    "                contig=contig,\n",
    "                filter_cm=filter_cm,\n",
    "                ) \n",
    "            for contig in tqdm(contigs)\n",
    "            ]\n",
    "    \n",
    "    if out_dir:\n",
    "        bkplt.output_file(filename=out_dir + cohort_id + \"_segments.html\", title=title)\n",
    "\n",
    "    fig = bklay.gridplot(\n",
    "        figs,\n",
    "        ncols=len(contigs),\n",
    "        toolbar_location=\"above\",\n",
    "        merge_tools=True,\n",
    "        height=height\n",
    "    ) \n",
    "    \n",
    "    if out_dir:\n",
    "        bkplt.save(fig)\n",
    "    \n",
    "    if show:\n",
    "        bkplt.show(fig)\n",
    "    \n",
    "    return fig\n",
    "    \n",
    "def _bokeh_style_genome_xaxis(fig, contig):\n",
    "    import bokeh.models as bkmod\n",
    "    \"\"\"Standard styling for X axis of genome plots.\"\"\"\n",
    "    fig.xaxis.axis_label = f\"Contig {contig} position (bp)\"\n",
    "    fig.xaxis.ticker = bkmod.AdaptiveTicker(min_interval=1)\n",
    "    fig.xaxis.minor_tick_line_color = None\n",
    "    fig.xaxis[0].formatter = bkmod.NumeralTickFormatter(format=\"0,0\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f24a8a1d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77547d97",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "481e2d85",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8286005",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "501ceb34",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# for cr in df_crosses.cross.unique()[:1]:\n",
    "#     df = df_crosses.query(\"cross == @cr\").dropna()\n",
    "#     father, mother = df.iloc[0, [2,3]]\n",
    "#     sibs = df.sample_id\n",
    "    \n",
    "#     df2  = df_ibd.query(\"iid1 in @sibs and iid2 in @sibs\")\n",
    "\n",
    "fig = plot_ibd_segments(\n",
    "        self=ag3, \n",
    "        df=df_ibd.sort_values('relationship'),\n",
    "        filter_cm=10,\n",
    "        out_dir=None,\n",
    "        show=True,\n",
    "        cohort_id='AG1000G-X',\n",
    "        contigs=('2RL', '3RL'), \n",
    "        height=1200\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "881eded4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
