{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "75d449b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import zarr\n",
    "import numpy as np\n",
    "import plotly.express as px"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "84bdcd65",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_zero_ibd_pairs(dfs_ibd, df_samples, fillna=0):\n",
    "\n",
    "    from itertools import combinations\n",
    "\n",
    "    dfs = [list(combinations(df_samples.query(\"sample_set == @s\")['sample_id'], 2)) for s in sample_sets]\n",
    "\n",
    "    comps = pd.concat([pd.DataFrame(d) for d in dfs], axis=0)\n",
    "    comps.columns = ['iid1', 'iid2']\n",
    "\n",
    "    def normalize_pairs(df, col1, col2):\n",
    "        df['normalized_iid1'] = df[[col1, col2]].min(axis=1)\n",
    "        df['normalized_iid2'] = df[[col1, col2]].max(axis=1)\n",
    "        return df\n",
    "\n",
    "    # Apply normalization\n",
    "    df_ibd_summary = normalize_pairs(dfs_ibd, 'iid1', 'iid2')\n",
    "    comps = normalize_pairs(comps, 'iid1', 'iid2')\n",
    "\n",
    "    df_ibd_summary = comps.merge(df_ibd_summary, how='left')\n",
    "\n",
    "    # Fill missing values with 0\n",
    "    df_ibd_summary.fillna({\n",
    "        'max_IBD': fillna,\n",
    "        'sum_IBD>8': 0,\n",
    "        'n_IBD>8': 0,\n",
    "        'sum_IBD>12': 0,\n",
    "        'n_IBD>12': 0,\n",
    "        'sum_IBD>16': 0,\n",
    "        'n_IBD>16': 0,\n",
    "        'sum_IBD>20': 0,\n",
    "    }, inplace=True)\n",
    "\n",
    "\n",
    "    df_ibd_summary = df_ibd_summary.merge(df_samples, left_on='iid1', right_on='sample_id').merge(df_samples, left_on='iid2', right_on='sample_id')\n",
    "    return df_ibd_summary"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd80b544",
   "metadata": {},
   "source": [
    "### Exploring IBD segments in ag3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3822f0d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7c1eba00",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_sets1 = pd.Series(glob.glob(\"../../ag3_results/ibd/*/ch2RL.tsv\")).str.split(\"/\").str.get(-2).to_list()\n",
    "sample_sets2 = pd.Series(glob.glob(\"../../ag3_results/ibd/*/ch3RL.tsv\")).str.split(\"/\").str.get(-2).to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "714ed409",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_sets = [item for item in sample_sets1 if item in sample_sets2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f2901efd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chromosome 2RL; Loaded 125 IBD\n",
      "Chromosome 3RL; Loaded 38 IBD\n",
      "Saved 163 IBD to ../../ag3_results/ibd/AG1000G-CM-C/ibd.tsv.\n",
      "Chromosome 2RL; Loaded 743 IBD\n",
      "Chromosome 3RL; Loaded 204 IBD\n",
      "Saved 947 IBD to ../../ag3_results/ibd/AG1000G-BF-A/ibd.tsv.\n",
      "Chromosome 2RL; Loaded 104 IBD\n",
      "Chromosome 3RL; Loaded 35 IBD\n",
      "Saved 139 IBD to ../../ag3_results/ibd/AG1000G-GN-A/ibd.tsv.\n",
      "Chromosome 2RL; Loaded 108 IBD\n",
      "Chromosome 3RL; Loaded 61 IBD\n",
      "Saved 169 IBD to ../../ag3_results/ibd/AG1000G-CF/ibd.tsv.\n",
      "Chromosome 2RL; Loaded 66 IBD\n",
      "Chromosome 3RL; Loaded 96 IBD\n",
      "Saved 162 IBD to ../../ag3_results/ibd/AG1000G-GM-C/ibd.tsv.\n",
      "Chromosome 2RL; Loaded 515 IBD\n",
      "Chromosome 3RL; Loaded 661 IBD\n",
      "Saved 1176 IBD to ../../ag3_results/ibd/AG1000G-MZ/ibd.tsv.\n",
      "Chromosome 2RL; Loaded 520 IBD\n",
      "Chromosome 3RL; Loaded 181 IBD\n",
      "Saved 701 IBD to ../../ag3_results/ibd/AG1000G-CD/ibd.tsv.\n",
      "Chromosome 2RL; Loaded 11 IBD\n",
      "Chromosome 3RL; Loaded 11 IBD\n",
      "Saved 22 IBD to ../../ag3_results/ibd/AG1000G-GQ/ibd.tsv.\n",
      "Chromosome 2RL; Loaded 68 IBD\n",
      "Chromosome 3RL; Loaded 18 IBD\n",
      "Saved 86 IBD to ../../ag3_results/ibd/AG1000G-ML-A/ibd.tsv.\n",
      "Chromosome 2RL; Loaded 11608 IBD\n",
      "Chromosome 3RL; Loaded 2479 IBD\n",
      "Saved 14087 IBD to ../../ag3_results/ibd/AG1000G-UG/ibd.tsv.\n",
      "Chromosome 2RL; Loaded 137 IBD\n",
      "Chromosome 3RL; Loaded 148 IBD\n",
      "Saved 285 IBD to ../../ag3_results/ibd/AG1000G-FR/ibd.tsv.\n",
      "Chromosome 2RL; Loaded 28 IBD\n",
      "Chromosome 3RL; Loaded 27 IBD\n",
      "Saved 55 IBD to ../../ag3_results/ibd/AG1000G-ML-B/ibd.tsv.\n",
      "Chromosome 2RL; Loaded 115 IBD\n",
      "Chromosome 3RL; Loaded 74 IBD\n",
      "Saved 189 IBD to ../../ag3_results/ibd/AG1000G-MW/ibd.tsv.\n",
      "Chromosome 2RL; Loaded 2 IBD\n",
      "Chromosome 3RL; Loaded 1 IBD\n",
      "Saved 3 IBD to ../../ag3_results/ibd/AG1000G-GM-B/ibd.tsv.\n",
      "Chromosome 2RL; Loaded 5838 IBD\n",
      "Chromosome 3RL; Loaded 5085 IBD\n",
      "Saved 10923 IBD to ../../ag3_results/ibd/AG1000G-KE/ibd.tsv.\n",
      "Chromosome 2RL; Loaded 1171 IBD\n",
      "Chromosome 3RL; Loaded 232 IBD\n",
      "Saved 1403 IBD to ../../ag3_results/ibd/AG1000G-AO/ibd.tsv.\n",
      "Chromosome 2RL; Loaded 212 IBD\n",
      "Chromosome 3RL; Loaded 61 IBD\n",
      "Saved 273 IBD to ../../ag3_results/ibd/AG1000G-BF-B/ibd.tsv.\n",
      "Chromosome 2RL; Loaded 143 IBD\n",
      "Chromosome 3RL; Loaded 174 IBD\n",
      "Saved 317 IBD to ../../ag3_results/ibd/AG1000G-GM-A/ibd.tsv.\n",
      "Chromosome 2RL; Loaded 44 IBD\n",
      "Chromosome 3RL; Loaded 26 IBD\n",
      "Saved 70 IBD to ../../ag3_results/ibd/AG1000G-GW/ibd.tsv.\n",
      "Chromosome 2RL; Loaded 189 IBD\n",
      "Chromosome 3RL; Loaded 160 IBD\n",
      "Saved 349 IBD to ../../ag3_results/ibd/AG1000G-GH/ibd.tsv.\n",
      "Chromosome 2RL; Loaded 12877 IBD\n",
      "Chromosome 3RL; Loaded 1755 IBD\n",
      "Saved 14632 IBD to ../../ag3_results/ibd/AG1000G-CM-A/ibd.tsv.\n",
      "Chromosome 2RL; Loaded 1119 IBD\n",
      "Chromosome 3RL; Loaded 673 IBD\n",
      "Saved 1792 IBD to ../../ag3_results/ibd/AG1000G-GA-A/ibd.tsv.\n",
      "Chromosome 2RL; Loaded 0 IBD\n",
      "Chromosome 3RL; Loaded 3 IBD\n",
      "Saved 3 IBD to ../../ag3_results/ibd/AG1000G-BF-C/ibd.tsv.\n",
      "Chromosome 2RL; Loaded 192 IBD\n",
      "Chromosome 3RL; Loaded 96 IBD\n",
      "Saved 288 IBD to ../../ag3_results/ibd/AG1000G-CM-B/ibd.tsv.\n",
      "Chromosome 2RL; Loaded 1362 IBD\n",
      "Chromosome 3RL; Loaded 502 IBD\n",
      "Saved 1864 IBD to ../../ag3_results/ibd/AG1000G-GN-B/ibd.tsv.\n",
      "Chromosome 2RL; Loaded 36 IBD\n",
      "Chromosome 3RL; Loaded 71 IBD\n",
      "Saved 107 IBD to ../../ag3_results/ibd/AG1000G-CI/ibd.tsv.\n",
      "Chromosome 2RL; Loaded 3126 IBD\n",
      "Chromosome 3RL; Loaded 2868 IBD\n",
      "Saved 5994 IBD to ../../ag3_results/ibd/AG1000G-TZ/ibd.tsv.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sanj/apps/mambaforge/lib/python3.11/site-packages/ancIBD/IO/ind_ibd.py:279: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  df_ibds = pd.concat(df_ibds)\n"
     ]
    }
   ],
   "source": [
    "from ancIBD.IO.ind_ibd import combine_all_chroms\n",
    "\n",
    "for sample_set in sample_sets:\n",
    "    combine_all_chroms(chs=('2RL', '3RL'),\n",
    "                       folder_base=f'../../ag3_results/ibd/{sample_set}/ch',\n",
    "                       path_save=f'../../ag3_results/ibd/{sample_set}/ibd.tsv'\n",
    "                      )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "945b79b2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> 8 cM: 163/163\n",
      "Of these with suff. SNPs per cM> 1000:               163/163\n",
      "ch\n",
      "2RL    125\n",
      "3RL     38\n",
      "Name: count, dtype: int64\n",
      "Saved 107 individual IBD pairs to: ../../ag3_results/ibd/AG1000G-CM-C/ibd_d220.tsv\n",
      "> 8 cM: 947/947\n",
      "Of these with suff. SNPs per cM> 1000:               947/947\n",
      "ch\n",
      "2RL    743\n",
      "3RL    204\n",
      "Name: count, dtype: int64\n",
      "Saved 904 individual IBD pairs to: ../../ag3_results/ibd/AG1000G-BF-A/ibd_d220.tsv\n",
      "> 8 cM: 139/139\n",
      "Of these with suff. SNPs per cM> 1000:               139/139\n",
      "ch\n",
      "2RL    104\n",
      "3RL     35\n",
      "Name: count, dtype: int64\n",
      "Saved 124 individual IBD pairs to: ../../ag3_results/ibd/AG1000G-GN-A/ibd_d220.tsv\n",
      "> 8 cM: 169/169\n",
      "Of these with suff. SNPs per cM> 1000:               169/169\n",
      "ch\n",
      "2RL    108\n",
      "3RL     61\n",
      "Name: count, dtype: int64\n",
      "Saved 159 individual IBD pairs to: ../../ag3_results/ibd/AG1000G-CF/ibd_d220.tsv\n",
      "> 8 cM: 162/162\n",
      "Of these with suff. SNPs per cM> 1000:               162/162\n",
      "ch\n",
      "3RL    96\n",
      "2RL    66\n",
      "Name: count, dtype: int64\n",
      "Saved 156 individual IBD pairs to: ../../ag3_results/ibd/AG1000G-GM-C/ibd_d220.tsv\n",
      "> 8 cM: 1176/1176\n",
      "Of these with suff. SNPs per cM> 1000:               1176/1176\n",
      "ch\n",
      "3RL    661\n",
      "2RL    515\n",
      "Name: count, dtype: int64\n",
      "Saved 983 individual IBD pairs to: ../../ag3_results/ibd/AG1000G-MZ/ibd_d220.tsv\n",
      "> 8 cM: 701/701\n",
      "Of these with suff. SNPs per cM> 1000:               701/701\n",
      "ch\n",
      "2RL    520\n",
      "3RL    181\n",
      "Name: count, dtype: int64\n",
      "Saved 594 individual IBD pairs to: ../../ag3_results/ibd/AG1000G-CD/ibd_d220.tsv\n",
      "> 8 cM: 22/22\n",
      "Of these with suff. SNPs per cM> 1000:               22/22\n",
      "ch\n",
      "2RL    11\n",
      "3RL    11\n",
      "Name: count, dtype: int64\n",
      "Saved 18 individual IBD pairs to: ../../ag3_results/ibd/AG1000G-GQ/ibd_d220.tsv\n",
      "> 8 cM: 86/86\n",
      "Of these with suff. SNPs per cM> 1000:               86/86\n",
      "ch\n",
      "2RL    68\n",
      "3RL    18\n",
      "Name: count, dtype: int64\n",
      "Saved 78 individual IBD pairs to: ../../ag3_results/ibd/AG1000G-ML-A/ibd_d220.tsv\n",
      "> 8 cM: 14087/14087\n",
      "Of these with suff. SNPs per cM> 1000:               14087/14087\n",
      "ch\n",
      "2RL    11608\n",
      "3RL     2479\n",
      "Name: count, dtype: int64\n",
      "Saved 12363 individual IBD pairs to: ../../ag3_results/ibd/AG1000G-UG/ibd_d220.tsv\n",
      "> 8 cM: 285/285\n",
      "Of these with suff. SNPs per cM> 1000:               285/285\n",
      "ch\n",
      "3RL    148\n",
      "2RL    137\n",
      "Name: count, dtype: int64\n",
      "Saved 189 individual IBD pairs to: ../../ag3_results/ibd/AG1000G-FR/ibd_d220.tsv\n",
      "> 8 cM: 55/55\n",
      "Of these with suff. SNPs per cM> 1000:               55/55\n",
      "ch\n",
      "2RL    28\n",
      "3RL    27\n",
      "Name: count, dtype: int64\n",
      "Saved 47 individual IBD pairs to: ../../ag3_results/ibd/AG1000G-ML-B/ibd_d220.tsv\n",
      "> 8 cM: 189/189\n",
      "Of these with suff. SNPs per cM> 1000:               189/189\n",
      "ch\n",
      "2RL    115\n",
      "3RL     74\n",
      "Name: count, dtype: int64\n",
      "Saved 177 individual IBD pairs to: ../../ag3_results/ibd/AG1000G-MW/ibd_d220.tsv\n",
      "> 8 cM: 3/3\n",
      "Of these with suff. SNPs per cM> 1000:               3/3\n",
      "ch\n",
      "2RL    2\n",
      "3RL    1\n",
      "Name: count, dtype: int64\n",
      "Saved 3 individual IBD pairs to: ../../ag3_results/ibd/AG1000G-GM-B/ibd_d220.tsv\n",
      "> 8 cM: 10923/10923\n",
      "Of these with suff. SNPs per cM> 1000:               10923/10923\n",
      "ch\n",
      "2RL    5838\n",
      "3RL    5085\n",
      "Name: count, dtype: int64\n",
      "Saved 1557 individual IBD pairs to: ../../ag3_results/ibd/AG1000G-KE/ibd_d220.tsv\n",
      "> 8 cM: 1403/1403\n",
      "Of these with suff. SNPs per cM> 1000:               1403/1403\n",
      "ch\n",
      "2RL    1171\n",
      "3RL     232\n",
      "Name: count, dtype: int64\n",
      "Saved 1245 individual IBD pairs to: ../../ag3_results/ibd/AG1000G-AO/ibd_d220.tsv\n",
      "> 8 cM: 273/273\n",
      "Of these with suff. SNPs per cM> 1000:               273/273\n",
      "ch\n",
      "2RL    212\n",
      "3RL     61\n",
      "Name: count, dtype: int64\n",
      "Saved 247 individual IBD pairs to: ../../ag3_results/ibd/AG1000G-BF-B/ibd_d220.tsv\n",
      "> 8 cM: 317/317\n",
      "Of these with suff. SNPs per cM> 1000:               317/317\n",
      "ch\n",
      "3RL    174\n",
      "2RL    143\n",
      "Name: count, dtype: int64\n",
      "Saved 296 individual IBD pairs to: ../../ag3_results/ibd/AG1000G-GM-A/ibd_d220.tsv\n",
      "> 8 cM: 70/70\n",
      "Of these with suff. SNPs per cM> 1000:               70/70\n",
      "ch\n",
      "2RL    44\n",
      "3RL    26\n",
      "Name: count, dtype: int64\n",
      "Saved 64 individual IBD pairs to: ../../ag3_results/ibd/AG1000G-GW/ibd_d220.tsv\n",
      "> 8 cM: 349/349\n",
      "Of these with suff. SNPs per cM> 1000:               349/349\n",
      "ch\n",
      "2RL    189\n",
      "3RL    160\n",
      "Name: count, dtype: int64\n",
      "Saved 299 individual IBD pairs to: ../../ag3_results/ibd/AG1000G-GH/ibd_d220.tsv\n",
      "> 8 cM: 14632/14632\n",
      "Of these with suff. SNPs per cM> 1000:               14632/14632\n",
      "ch\n",
      "2RL    12877\n",
      "3RL     1755\n",
      "Name: count, dtype: int64\n",
      "Saved 12905 individual IBD pairs to: ../../ag3_results/ibd/AG1000G-CM-A/ibd_d220.tsv\n",
      "> 8 cM: 1792/1792\n",
      "Of these with suff. SNPs per cM> 1000:               1792/1792\n",
      "ch\n",
      "2RL    1119\n",
      "3RL     673\n",
      "Name: count, dtype: int64\n",
      "Saved 1290 individual IBD pairs to: ../../ag3_results/ibd/AG1000G-GA-A/ibd_d220.tsv\n",
      "> 8 cM: 3/3\n",
      "Of these with suff. SNPs per cM> 1000:               3/3\n",
      "ch\n",
      "3RL    3\n",
      "Name: count, dtype: int64\n",
      "Saved 2 individual IBD pairs to: ../../ag3_results/ibd/AG1000G-BF-C/ibd_d220.tsv\n",
      "> 8 cM: 288/288\n",
      "Of these with suff. SNPs per cM> 1000:               288/288\n",
      "ch\n",
      "2RL    192\n",
      "3RL     96\n",
      "Name: count, dtype: int64\n",
      "Saved 256 individual IBD pairs to: ../../ag3_results/ibd/AG1000G-CM-B/ibd_d220.tsv\n",
      "> 8 cM: 1864/1864\n",
      "Of these with suff. SNPs per cM> 1000:               1864/1864\n",
      "ch\n",
      "2RL    1362\n",
      "3RL     502\n",
      "Name: count, dtype: int64\n",
      "Saved 1659 individual IBD pairs to: ../../ag3_results/ibd/AG1000G-GN-B/ibd_d220.tsv\n",
      "> 8 cM: 107/107\n",
      "Of these with suff. SNPs per cM> 1000:               107/107\n",
      "ch\n",
      "3RL    71\n",
      "2RL    36\n",
      "Name: count, dtype: int64\n",
      "Saved 107 individual IBD pairs to: ../../ag3_results/ibd/AG1000G-CI/ibd_d220.tsv\n",
      "> 8 cM: 5994/5994\n",
      "Of these with suff. SNPs per cM> 1000:               5994/5994\n",
      "ch\n",
      "2RL    3126\n",
      "3RL    2868\n",
      "Name: count, dtype: int64\n",
      "Saved 4980 individual IBD pairs to: ../../ag3_results/ibd/AG1000G-TZ/ibd_d220.tsv\n"
     ]
    }
   ],
   "source": [
    "from ancIBD.IO.ind_ibd import create_ind_ibd_df\n",
    "\n",
    "dfs = []\n",
    "for sample_set in sample_sets:\n",
    "    dfs.append(create_ind_ibd_df(ibd_data = f'../../ag3_results/ibd/{sample_set}/ibd.tsv',\n",
    "                          min_cms = [8, 12, 16, 20], snp_cm = 1000, min_cm = 8, sort_col = 0,\n",
    "                          savepath = f'../../ag3_results/ibd/{sample_set}/ibd_d220.tsv').assign(sample_set=sample_set\n",
    "                                                                                            ))\n",
    "    \n",
    "dfs_ibd = pd.concat(dfs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4eac2d5a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  const force = true;\n",
       "\n",
       "  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "const JS_MIME_TYPE = 'application/javascript';\n",
       "  const HTML_MIME_TYPE = 'text/html';\n",
       "  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  const CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    const script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    function drop(id) {\n",
       "      const view = Bokeh.index.get_by_id(id)\n",
       "      if (view != null) {\n",
       "        view.model.document.clear()\n",
       "        Bokeh.index.delete(view)\n",
       "      }\n",
       "    }\n",
       "\n",
       "    const cell = handle.cell;\n",
       "\n",
       "    const id = cell.output_area._bokeh_element_id;\n",
       "    const server_id = cell.output_area._bokeh_server_id;\n",
       "\n",
       "    // Clean up Bokeh references\n",
       "    if (id != null) {\n",
       "      drop(id)\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      const cmd_clean = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd_clean, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            const id = msg.content.text.trim()\n",
       "            drop(id)\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      const cmd_destroy = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd_destroy);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    const output_area = handle.output_area;\n",
       "    const output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    const toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      const bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      const script_attrs = bk_div.children[0].attributes;\n",
       "      for (let i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      const toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    const events = require('base/js/events');\n",
       "    const OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  const NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded() {\n",
       "    const el = document.getElementById(null);\n",
       "    if (el != null) {\n",
       "      el.textContent = \"BokehJS is loading...\";\n",
       "    }\n",
       "    if (root.Bokeh !== undefined) {\n",
       "      if (el != null) {\n",
       "        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(display_loaded, 100)\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) {\n",
       "        if (callback != null)\n",
       "          callback();\n",
       "      });\n",
       "    } finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.debug(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(css_urls, js_urls, callback) {\n",
       "    if (css_urls == null) css_urls = [];\n",
       "    if (js_urls == null) js_urls = [];\n",
       "\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = css_urls.length + js_urls.length;\n",
       "\n",
       "    function on_load() {\n",
       "      root._bokeh_is_loading--;\n",
       "      if (root._bokeh_is_loading === 0) {\n",
       "        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n",
       "        run_callbacks()\n",
       "      }\n",
       "    }\n",
       "\n",
       "    function on_error(url) {\n",
       "      console.error(\"failed to load \" + url);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < css_urls.length; i++) {\n",
       "      const url = css_urls[i];\n",
       "      const element = document.createElement(\"link\");\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.rel = \"stylesheet\";\n",
       "      element.type = \"text/css\";\n",
       "      element.href = url;\n",
       "      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n",
       "      document.body.appendChild(element);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < js_urls.length; i++) {\n",
       "      const url = js_urls[i];\n",
       "      const element = document.createElement('script');\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.async = false;\n",
       "      element.src = url;\n",
       "      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.head.appendChild(element);\n",
       "    }\n",
       "  };\n",
       "\n",
       "  function inject_raw_css(css) {\n",
       "    const element = document.createElement(\"style\");\n",
       "    element.appendChild(document.createTextNode(css));\n",
       "    document.body.appendChild(element);\n",
       "  }\n",
       "\n",
       "  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-3.3.4.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.3.4.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.3.4.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.3.4.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.3.4.min.js\"];\n",
       "  const css_urls = [];\n",
       "\n",
       "  const inline_js = [    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "function(Bokeh) {\n",
       "    }\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    if (root.Bokeh !== undefined || force === true) {\n",
       "          for (let i = 0; i < inline_js.length; i++) {\n",
       "      inline_js[i].call(root, root.Bokeh);\n",
       "    }\n",
       "} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      const cell = $(document.getElementById(null)).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(css_urls, js_urls, function() {\n",
       "      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  const force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n\n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  const NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    const el = document.getElementById(null);\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-3.3.4.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.3.4.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.3.4.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.3.4.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.3.4.min.js\"];\n  const css_urls = [];\n\n  const inline_js = [    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\nfunction(Bokeh) {\n    }\n  ];\n\n  function run_inline_js() {\n    if (root.Bokeh !== undefined || force === true) {\n          for (let i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\n} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      const cell = $(document.getElementById(null)).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                     \r"
     ]
    }
   ],
   "source": [
    "import malariagen_data\n",
    "import matplotlib.pyplot as plt\n",
    "ag3 = malariagen_data.Ag3()\n",
    "df_samples = ag3.sample_metadata(sample_sets)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81c4cc93",
   "metadata": {},
   "source": [
    "Check that we now have the correct number of pairwise comparisons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e41a35b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_samples = df_samples.assign(set_taxon=lambda x: x.sample_set + \"-\" + x.taxon)\n",
    "\n",
    "from itertools import combinations\n",
    "\n",
    "# add zero ibd pairs\n",
    "df_ibd_summary = add_zero_ibd_pairs(dfs_ibd, df_samples, fillna=0).rename(columns={'max_IBD':'total_ibd_cm'})\n",
    "# assign a taxon + sample set column\n",
    "df_ibd_summary = df_ibd_summary.assign(set_taxon=[y if y == df_ibd_summary.set_taxon_x[i] else 'between-species' for i,y in enumerate(df_ibd_summary.set_taxon_y)])\n",
    "# remove combos with few observations\n",
    "df_ibd_summary = df_ibd_summary.groupby('set_taxon').filter(lambda g: len(g) >= 100)\n",
    "\n",
    "assert df_ibd_summary.query(\"sample_set == 'AG1000G-GH'\").shape[0] == len(list(combinations(df_samples.query(\"sample_set == 'AG1000G-GH'\").sample_id, 2)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d39d745",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"plotting\")\n",
    "fig = px.box(\n",
    "    df_ibd_summary.sort_values('set_taxon'),\n",
    "    y='total_ibd_cm', \n",
    "    x='set_taxon',\n",
    "    template='simple_white', \n",
    "    color='country_x', \n",
    "    title='total shared IBD within sample sets / taxon combo',\n",
    "    height=700, \n",
    "    width=1000\n",
    ")\n",
    "fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b753670",
   "metadata": {},
   "outputs": [],
   "source": [
    "ibd_rate = df_ibd_summary.groupby('set_taxon').agg({'total_ibd_cm':'sum'}).assign(ibd_rate=lambda x: x.total_ibd_cm / df_ibd_summary.groupby('set_taxon').size())\n",
    "ibd_rate = ibd_rate.reset_index()\n",
    "\n",
    "fig = px.bar(\n",
    "    ibd_rate.sort_values('set_taxon'),\n",
    "    y='ibd_rate', \n",
    "    x='set_taxon',\n",
    "    template='simple_white', \n",
    "    title='Distribution of total shared IBD between individuals in ag3',\n",
    "    height=500, \n",
    "    width=800\n",
    ")\n",
    "fig"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6563a88a",
   "metadata": {},
   "source": [
    "### Plottijng with bokeh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92364123",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_ibd_segments_contig(\n",
    "    self,\n",
    "    df,\n",
    "    contig, \n",
    "    filter_cm=None,\n",
    "    show=False,\n",
    "    width=None,\n",
    "    height=400\n",
    "):\n",
    "    from itertools import combinations\n",
    "    import bokeh.plotting as bkplt\n",
    "    import bokeh.models as bkmod\n",
    "    import bokeh.layouts as bklay\n",
    "    \n",
    "    df = df.query(\"ch == @contig\")\n",
    "    \n",
    "    if filter_cm:\n",
    "        df = df.query(\"cm > @filter_cm\")\n",
    "        \n",
    "    print(\"finding levels\")\n",
    "    level_df = df[['iid1', 'iid2']].drop_duplicates()\n",
    "    # Add the 'level' column with values from 0 to 1\n",
    "    level_df['level'] = np.linspace(0, 1, level_df.shape[0])\n",
    "    # # Join with the original dataframe\n",
    "    df = df.merge(level_df, on=['iid1', 'iid2'], how='left')\n",
    "    \n",
    "#     print(\"colour mapping\")\n",
    "#     colour_mapping = {'half-ibd':'gray', \n",
    "#                       'full-ibd':'blue'}\n",
    "#     colour = df['ibd_type'].apply(lambda x: colour_mapping[x])\n",
    "\n",
    "    source = bkmod.ColumnDataSource(data={\n",
    "         'sample_id1':df['iid1'].to_numpy(),\n",
    "         'sample_id2':df['iid2'].to_numpy(),\n",
    "        'chromosome': df['ch'].to_numpy(),\n",
    "        'start': df['StartBP'].to_numpy(),\n",
    "        'end': df['EndBP'].to_numpy(),\n",
    "        'cm_length': df['lengthM'].to_numpy(),\n",
    "        'bottoms':df['level'].to_numpy(),\n",
    "        'tops': df['level'].to_numpy()+0.0001,\n",
    "#         'colour': colour.to_numpy()\n",
    "    })\n",
    "\n",
    "    hover = bkmod.HoverTool(tooltips=[\n",
    "             (\"sample_id1\", '@sample_id1'),\n",
    "             (\"sample_id2\", '@sample_id2'),\n",
    "             ('cm_length', '@cm_length'),\n",
    "            (\"segment span\", \"@start{,} - @end{,}\"),\n",
    "        ])\n",
    "        \n",
    "    print(\"making figure\")\n",
    "    if not width:\n",
    "        width = int(self.genome_sequence(contig).shape[0]/200000)\n",
    "    fig1 = bkplt.figure(title=contig,\n",
    "                        width=width,\n",
    "                        height=500, \n",
    "                        tools=\"tap,box_zoom,xpan,xzoom_in,xzoom_out,xwheel_zoom,reset\".split() + [hover],\n",
    "                        toolbar_location='above', active_drag='xpan', active_scroll='xwheel_zoom',\n",
    "                        output_backend=\"webgl\")\n",
    "\n",
    "    glyph = bkmod.Quad(left='start', right='end', bottom='bottoms', top='tops', line_color=\"grey\", line_alpha=.8, line_width=1)\n",
    "    fig1.add_glyph(source, glyph)\n",
    "\n",
    "    fig1.x_range = bkmod.Range1d(0, self.genome_sequence(contig).shape[0], bounds='auto')\n",
    "    fig1.y_range = bkmod.Range1d(0, 1, bounds='auto')\n",
    "    fig1.x_range.max_interval = self.genome_sequence(contig).shape[0]\n",
    "    fig1.yaxis.visible = False\n",
    "    fig1.ygrid.visible = False\n",
    "    _bokeh_style_genome_xaxis(fig1, contig)\n",
    "    \n",
    "    if show:\n",
    "        bkplt.show(fig1)\n",
    "    \n",
    "    return fig1\n",
    "\n",
    "def plot_ibd_segments(\n",
    "        self,\n",
    "        df, \n",
    "        out_dir,\n",
    "        cohort_id,\n",
    "        contigs=('2RL', '3RL', 'X'),\n",
    "        filter_cm=None,\n",
    "        show=True,\n",
    "        title=None,\n",
    "        height=1000\n",
    "    ):\n",
    "    import bokeh.models as bkmod\n",
    "    import bokeh.layouts as bklay\n",
    "    import bokeh.plotting as bkplt\n",
    "    from tqdm.notebook import tqdm\n",
    "    \n",
    "    figs = [\n",
    "            plot_ibd_segments_contig(\n",
    "                self=self,\n",
    "                df=df,\n",
    "                contig=contig,\n",
    "                filter_cm=filter_cm,\n",
    "                ) \n",
    "            for contig in tqdm(contigs)\n",
    "            ]\n",
    "    \n",
    "    if out_dir:\n",
    "        bkplt.output_file(filename=out_dir + cohort_id + \"_segments.html\", title=title)\n",
    "\n",
    "    fig = bklay.gridplot(\n",
    "        figs,\n",
    "        ncols=len(contigs),\n",
    "        toolbar_location=\"above\",\n",
    "        merge_tools=True,\n",
    "        height=height\n",
    "    ) \n",
    "    \n",
    "    if out_dir:\n",
    "        bkplt.save(fig)\n",
    "    \n",
    "    if show:\n",
    "        bkplt.show(fig)\n",
    "    \n",
    "    return fig\n",
    "    \n",
    "def _bokeh_style_genome_xaxis(fig, contig):\n",
    "    import bokeh.models as bkmod\n",
    "    \"\"\"Standard styling for X axis of genome plots.\"\"\"\n",
    "    fig.xaxis.axis_label = f\"Contig {contig} position (bp)\"\n",
    "    fig.xaxis.ticker = bkmod.AdaptiveTicker(min_interval=1)\n",
    "    fig.xaxis.minor_tick_line_color = None\n",
    "    fig.xaxis[0].formatter = bkmod.NumeralTickFormatter(format=\"0,0\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f24a8a1d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77547d97",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "481e2d85",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8286005",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "501ceb34",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for cr in df_crosses.cross.unique()[:1]:\n",
    "#     df = df_crosses.query(\"cross == @cr\").dropna()\n",
    "#     father, mother = df.iloc[0, [2,3]]\n",
    "#     sibs = df.sample_id\n",
    "    \n",
    "#     df2  = df_ibd.query(\"iid1 in @sibs and iid2 in @sibs\")\n",
    "\n",
    "fig = plot_ibd_segments(\n",
    "        self=ag3, \n",
    "        df=df_ibd.sort_values('relationship'),\n",
    "        filter_cm=10,\n",
    "        out_dir=None,\n",
    "        show=True,\n",
    "        cohort_id='AG1000G-X',\n",
    "        contigs=('2RL', '3RL'), \n",
    "        height=1200\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "881eded4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
