{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import allel\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import zarr\n",
    "from pathlib import Path\n",
    "import scipy\n",
    "import seaborn as sns\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "%run tools.py"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GAARDIAN - Ghana - G12"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_set = \"1244-VO-GH-YAWSON-VMF00149\"\n",
    "chroms = ['2L', '2R', '3L', '3R', 'X']\n",
    "\n",
    "snps = {}\n",
    "pos = {}\n",
    "\n",
    "for chrom in chroms:\n",
    "    \n",
    "    snps[chrom], pos[chrom], metadata = load_arrays_and_metadata(\"../resources\", sample_set, chrom)\n",
    "    \n",
    "metadata['location2'] = metadata['location'].str.split(\".\").str.get(0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Lets load the sample metadata and the SNP genotypes and positions. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "my_sample_set = '1244-VO-GH-YAWSON-VMF00149'\n",
    "metadata = ag3.sample_metadata(sample_sets=my_sample_set)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Lets make a new column in the metadata of overall village, rather than village x house. We split on full stops and take the first element. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata['location2'] = metadata['location'].str.split(\".\").str.get(0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How many do we have from each species?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "coluzzii    422\n",
       "gambiae      63\n",
       "Name: species, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata.species.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How many samples do we have from each village??"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Domenase         57\n",
       "Annorkrom        49\n",
       "New Edubiase     42\n",
       "Pokukrom         37\n",
       "Odumto           34\n",
       "Koniyaw          32\n",
       "Nkotumso         30\n",
       "Kokotro          28\n",
       "Prisintease      24\n",
       "Kente            24\n",
       "Watreso          19\n",
       "Adansi Apagya    16\n",
       "Wamase           16\n",
       "Ankaako          12\n",
       "Mprakyire        10\n",
       "Anwona            8\n",
       "Adumanu           6\n",
       "Yadome            6\n",
       "Adansi-Krom       6\n",
       "Bogyawe           5\n",
       "Asonkore          5\n",
       "Jacobu            5\n",
       "Fumso             4\n",
       "Dompoase          3\n",
       "Anhwiaso          2\n",
       "Subin Camp        2\n",
       "Bogobiri West     2\n",
       "Mensonso          1\n",
       "Name: location2, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata.location2.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each location, how many gambiae / coluzzii do we have?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>species</th>\n",
       "      <th>coluzzii</th>\n",
       "      <th>gambiae</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>location2</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Adansi Apagya</th>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Adansi-Krom</th>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Adumanu</th>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Anhwiaso</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ankaako</th>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Annorkrom</th>\n",
       "      <td>46</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Anwona</th>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Asonkore</th>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bogobiri West</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bogyawe</th>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Domenase</th>\n",
       "      <td>57</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dompoase</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Fumso</th>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Jacobu</th>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Kente</th>\n",
       "      <td>23</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Kokotro</th>\n",
       "      <td>24</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Koniyaw</th>\n",
       "      <td>31</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mensonso</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mprakyire</th>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>New Edubiase</th>\n",
       "      <td>15</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Nkotumso</th>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Odumto</th>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pokukrom</th>\n",
       "      <td>37</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Prisintease</th>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Subin Camp</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Wamase</th>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Watreso</th>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Yadome</th>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "species        coluzzii  gambiae\n",
       "location2                       \n",
       "Adansi Apagya        15        1\n",
       "Adansi-Krom           6        0\n",
       "Adumanu               5        1\n",
       "Anhwiaso              2        0\n",
       "Ankaako              12        0\n",
       "Annorkrom            46        3\n",
       "Anwona                5        3\n",
       "Asonkore              0        5\n",
       "Bogobiri West         0        2\n",
       "Bogyawe               3        2\n",
       "Domenase             57        0\n",
       "Dompoase              1        2\n",
       "Fumso                 1        3\n",
       "Jacobu                2        3\n",
       "Kente                23        1\n",
       "Kokotro              24        4\n",
       "Koniyaw              31        1\n",
       "Mensonso              0        1\n",
       "Mprakyire             9        1\n",
       "New Edubiase         15       27\n",
       "Nkotumso             30        0\n",
       "Odumto               32        2\n",
       "Pokukrom             37        0\n",
       "Prisintease          24        0\n",
       "Subin Camp            1        1\n",
       "Wamase               16        0\n",
       "Watreso              19        0\n",
       "Yadome                6        0"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.crosstab(metadata.location2, metadata.species)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And what does the dask SNP array look like?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## G12"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cluster_G12(gnalt, cut_height=0.1, metric='euclidean'):\n",
    "    \"\"\"\n",
    "    Hierarchically clusters genotypes and calculates G12 statistic. \n",
    "    \"\"\"\n",
    "    # cluster the genotypes in the window\n",
    "    dist = scipy.spatial.distance.pdist(gnalt.T, metric=metric)\n",
    "    if metric in {'hamming', 'jaccard'}:\n",
    "        # convert distance to number of SNPs, easier to interpret\n",
    "        dist *= gnalt.shape[0]\n",
    "\n",
    "    Z = scipy.cluster.hierarchy.linkage(dist, method='single')\n",
    "    cut = scipy.cluster.hierarchy.cut_tree(Z, height=cut_height)[:, 0]\n",
    "    cluster_sizes = np.bincount(cut)\n",
    "    clusters = [np.nonzero(cut == i)[0] for i in range(cut.max() + 1)]\n",
    "    \n",
    "    # get freq of clusters and sort by largest freq\n",
    "    freqs = cluster_sizes/gnalt.shape[1]\n",
    "    freqs = np.sort(freqs)[::-1]\n",
    "    \n",
    "    # calculate g12\n",
    "    g12 = np.sum(freqs[:2])**2 + np.sum(freqs[2:]**2)\n",
    "    \n",
    "    return(g12)\n",
    "\n",
    "def garuds_G12(gnalt, pos, cut_height=None, save=False, prefix=None, name=None, metric='euclidean', window_size=1000, step_size=500, cluster=False, species=None):\n",
    "    \n",
    "    \"\"\"\n",
    "    Calculates G12, stores windows in .tsv and plots\n",
    "    \"\"\"\n",
    "        \n",
    "    # Do we want to cluster the Multi-locus genotypes (MLGs), or just group MLGs if they are identical\n",
    "    if cluster:\n",
    "        g12 = allel.moving_statistic(gnalt, cluster_G12, size=window_size, step=step_size, metric=metric, cut_height=cut_height)\n",
    "    else:\n",
    "        g12,_,_,_ = allel.moving_garud_h(gnalt, size=window_size, step=step_size)\n",
    "\n",
    "    midpoint = allel.moving_statistic(pos, np.median, size=window_size, step=step_size)\n",
    "    \n",
    "    # store windowed G12 statistics as .tsv \n",
    "    df = pd.DataFrame({'midpoint':midpoint, 'G12':g12})\n",
    "    df.to_csv(f\"{prefix}/{name}.tsv\", sep=\"\\t\", index=False)\n",
    "    \n",
    "    # create plot \n",
    "    xtick = np.arange(0, midpoint.max(), 2000000)\n",
    "    ylim = np.max([0.50, g12.max()])\n",
    "    plt.figure(figsize=[20,10])\n",
    "    if species == 'gambiae':\n",
    "        sns.lineplot(midpoint, g12, color='lightcoral')\n",
    "    else:\n",
    "        sns.lineplot(midpoint, g12, color='dodgerblue')\n",
    "    plt.ylim(0, ylim)\n",
    "    plt.xticks(xtick, rotation=45, ha='right')\n",
    "    plt.ticklabel_format(style='plain', axis='x')\n",
    "    plt.title(f\"G12 {name}\")\n",
    "    if save: plt.savefig(f\"{prefix}/{name}.png\",format=\"png\")\n",
    "    \n",
    "    plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run G12"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "chroms = ['2L', '2R', '3R', '3L', 'X']\n",
    "chrom = '2L'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "skipping Kente_coluzzii_2L, as already done.\n",
      "skipping Odumto_coluzzii_2L, as already done.\n",
      "skipping Watreso_coluzzii_2L, as already done.\n",
      "skipping Koniyaw_coluzzii_2L, as already done.\n",
      "skipping Kokotro_coluzzii_2L, as already done.\n",
      "skipping New Edubiase_gambiae_2L, as already done.\n",
      "skipping New Edubiase_coluzzii_2L, as already done.\n",
      "skipping Adansi Apagya_coluzzii_2L, as already done.\n",
      "skipping Wamase_coluzzii_2L, as already done.\n",
      "skipping Annorkrom_coluzzii_2L, as already done.\n",
      "skipping Prisintease_coluzzii_2L, as already done.\n",
      "skipping Pokukrom_coluzzii_2L, as already done.\n",
      "skipping Nkotumso_coluzzii_2L, as already done.\n",
      "skipping Domenase_coluzzii_2L, as already done.\n",
      "skipping Kente_coluzzii_2R, as already done.\n",
      "skipping Odumto_coluzzii_2R, as already done.\n",
      "skipping Watreso_coluzzii_2R, as already done.\n",
      "skipping Koniyaw_coluzzii_2R, as already done.\n",
      "skipping Kokotro_coluzzii_2R, as already done.\n",
      "skipping New Edubiase_gambiae_2R, as already done.\n",
      "skipping New Edubiase_coluzzii_2R, as already done.\n",
      "skipping Adansi Apagya_coluzzii_2R, as already done.\n",
      "skipping Wamase_coluzzii_2R, as already done.\n",
      "skipping Annorkrom_coluzzii_2R, as already done.\n",
      "skipping Prisintease_coluzzii_2R, as already done.\n",
      "skipping Pokukrom_coluzzii_2R, as already done.\n",
      "skipping Nkotumso_coluzzii_2R, as already done.\n",
      "skipping Domenase_coluzzii_2R, as already done.\n",
      "skipping Kente_coluzzii_3R, as already done.\n",
      "skipping Odumto_coluzzii_3R, as already done.\n",
      "skipping Watreso_coluzzii_3R, as already done.\n",
      "skipping Koniyaw_coluzzii_3R, as already done.\n",
      "skipping Kokotro_coluzzii_3R, as already done.\n",
      "skipping New Edubiase_gambiae_3R, as already done.\n",
      "skipping New Edubiase_coluzzii_3R, as already done.\n",
      "skipping Adansi Apagya_coluzzii_3R, as already done.\n",
      "skipping Wamase_coluzzii_3R, as already done.\n",
      "skipping Annorkrom_coluzzii_3R, as already done.\n",
      "skipping Prisintease_coluzzii_3R, as already done.\n",
      "skipping Pokukrom_coluzzii_3R, as already done.\n",
      "skipping Nkotumso_coluzzii_3R, as already done.\n",
      "skipping Domenase_coluzzii_3R, as already done.\n",
      "skipping Kente_coluzzii_3L, as already done.\n",
      "skipping Odumto_coluzzii_3L, as already done.\n",
      "skipping Watreso_coluzzii_3L, as already done.\n",
      "skipping Koniyaw_coluzzii_3L, as already done.\n",
      "skipping Kokotro_coluzzii_3L, as already done.\n",
      "skipping New Edubiase_gambiae_3L, as already done.\n",
      "skipping New Edubiase_coluzzii_3L, as already done.\n",
      "skipping Adansi Apagya_coluzzii_3L, as already done.\n",
      "skipping Wamase_coluzzii_3L, as already done.\n",
      "skipping Annorkrom_coluzzii_3L, as already done.\n",
      "skipping Prisintease_coluzzii_3L, as already done.\n",
      "skipping Pokukrom_coluzzii_3L, as already done.\n",
      "skipping Nkotumso_coluzzii_3L, as already done.\n",
      "skipping Domenase_coluzzii_3L, as already done.\n",
      "skipping Kente_coluzzii_X, as already done.\n",
      "skipping Odumto_coluzzii_X, as already done.\n",
      "skipping Watreso_coluzzii_X, as already done.\n",
      "skipping Koniyaw_coluzzii_X, as already done.\n",
      "skipping Kokotro_coluzzii_X, as already done.\n",
      "skipping New Edubiase_gambiae_X, as already done.\n",
      "skipping New Edubiase_coluzzii_X, as already done.\n",
      "skipping Adansi Apagya_coluzzii_X, as already done.\n",
      "skipping Wamase_coluzzii_X, as already done.\n",
      "skipping Annorkrom_coluzzii_X, as already done.\n",
      "skipping Prisintease_coluzzii_X, as already done.\n",
      "skipping Pokukrom_coluzzii_X, as already done.\n",
      "skipping Nkotumso_coluzzii_X, as already done.\n",
      "skipping Domenase_coluzzii_X, as already done.\n"
     ]
    }
   ],
   "source": [
    "for chrom in chroms:\n",
    "    \n",
    "    # Load site filters, SNP and POS arrays and filter\n",
    "    gambcolu_filter = ag3.site_filters(mask=\"gamb_colu\", contig=chrom)\n",
    "    snps = allel.GenotypeDaskArray(ag3.snp_genotypes(contig=chrom, sample_sets=my_sample_set))\n",
    "    snps = snps.compress(gambcolu_filter, axis=0)\n",
    "    pos = ag3.snp_sites(contig=chrom)\n",
    "    pos = allel.SortedIndex(pos[0])[gambcolu_filter]\n",
    "\n",
    "    # Count alleles per site\n",
    "    ac = snps.count_alleles()\n",
    "    # Get boolean array for biallelic or not\n",
    "    bial_bool = ac.is_biallelic()\n",
    "    # Filter arrays to biallelic only\n",
    "    snps = snps.compress(bial_bool, axis=0)\n",
    "    positions = pos[bial_bool]\n",
    "\n",
    "    for loc in metadata.location2.unique():\n",
    "\n",
    "        nmeta = metadata[metadata.location2 == loc]\n",
    "\n",
    "            # have edited .species_gambiae_coluzzii column to contain 'arabiensis' instead of NA \n",
    "        for sp in nmeta.species_gambiae_coluzzii.unique():\n",
    "\n",
    "            # if there is less than n samples then skip\n",
    "            if (nmeta.species_gambiae_coluzzii == sp).sum() <= 12:\n",
    "                continue\n",
    "                \n",
    "            # if file exists ignore and skip\n",
    "            myfile = Path(f\"gaardian/{loc}_{sp}_{chrom}.G12.png\")\n",
    "            if myfile.is_file():\n",
    "                print(f\"skipping {loc}_{sp}_{chrom}, as already done.\")\n",
    "                continue\n",
    "\n",
    "            #filter to species \n",
    "            nmeta3 = nmeta[nmeta.species_gambiae_coluzzii == sp]\n",
    "            flt = np.array(nmeta3.index)\n",
    "            # filter to correct loc, year, species individuals\n",
    "            gn = snps.take(flt, axis=1)\n",
    "\n",
    "            print(f\"--------- Running G12 on {loc} {sp} Chromosome {chrom} ----------\")\n",
    "\n",
    "            # filter to segregating sites only as the phase 3 arrays contain every SNP on the chromosome\n",
    "            seg = gn.count_alleles().is_segregating()\n",
    "            gn = gn.compress(seg, axis=0)\n",
    "            pos = positions[seg]\n",
    "\n",
    "            print(f\"There are {gn.shape[0]} segregating sites and {gn.shape[1]} individuals\")\n",
    "\n",
    "            # convert to 0,1,2 genotypes\n",
    "            gn_alt = gn.to_n_alt().compute()\n",
    "\n",
    "            # calculate G12 and plot figs \n",
    "            garuds_G12(gnalt=gn_alt, pos=pos, cut_height=6,\n",
    "                       save=True, \n",
    "                       prefix='gaardian',\n",
    "                       name=f\"{loc}_{sp}_{chrom}.G12\", \n",
    "                       metric='euclidean',\n",
    "                      window_size=1000,\n",
    "                      step_size=500, cluster=True, species=sp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.7.3 64-bit ('pysanj': conda)",
   "language": "python",
   "name": "python37364bitpysanjconda39958c5ba9ef41c3a5542f2bfe81dc48"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
