{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import malariagen_data\n",
    "import allel\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import dask\n",
    "from dask.diagnostics import ProgressBar\n",
    "import zarr\n",
    "from pathlib import Path\n",
    "import scipy\n",
    "import seaborn as sns\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "%run tools.py"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Karyotyping the GAARDIAN data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "ag3 = malariagen_data.Ag3(\"gs://vo_agam_release/\", pre=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_set</th>\n",
       "      <th>sample_count</th>\n",
       "      <th>release</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1191-VO-MULTI-OLOUGHLIN-VMF00106</td>\n",
       "      <td>237</td>\n",
       "      <td>v3.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1191-VO-MULTI-OLOUGHLIN-VMF00140</td>\n",
       "      <td>1095</td>\n",
       "      <td>v3.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1244-VO-GH-YAWSON-VMF00149</td>\n",
       "      <td>485</td>\n",
       "      <td>v3.4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         sample_set  sample_count release\n",
       "0  1191-VO-MULTI-OLOUGHLIN-VMF00106           237    v3.4\n",
       "1  1191-VO-MULTI-OLOUGHLIN-VMF00140          1095    v3.4\n",
       "2        1244-VO-GH-YAWSON-VMF00149           485    v3.4"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sample_sets = ag3.sample_sets(release=\"v3.4\")\n",
    "df_sample_sets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "my_sample_set = '1244-VO-GH-YAWSON-VMF00149'\n",
    "metadata = ag3.sample_metadata(sample_sets=my_sample_set)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata['location2'] = metadata['location'].str.split(\".\").str.get(0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## compkaryo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- Running CompKaryo 2La--- \n",
      "--- Running CompKaryo 2Rj--- \n",
      "--- Running CompKaryo 2Rb--- \n",
      "--- Running CompKaryo 2Rc_col--- \n",
      "--- Running CompKaryo 2Rc_gam--- \n",
      "--- Running CompKaryo 2Rd--- \n",
      "--- Running CompKaryo 2Ru--- \n"
     ]
    }
   ],
   "source": [
    "invDict = {}\n",
    "\n",
    "for inversion in inversionDict.keys():\n",
    "    chrom = inversionDict[inversion][0]\n",
    "\n",
    "    snps = allel.GenotypeDaskArray(ag3.snp_genotypes(contig=chrom, sample_sets=my_sample_set))\n",
    "    pos = ag3.snp_sites(contig=chrom)\n",
    "    pos = allel.SortedIndex(pos[0])\n",
    "\n",
    "    callset = {'geno':snps, 'pos':pos, 'chrom':chrom}\n",
    "    print(f\"--- Running CompKaryo {inversion}--- \")\n",
    "    av_gts, total_sites, num_0, num_1, num_2 = compkaryo(callset, inversion)\n",
    "    total_sites = total_sites.compute()\n",
    "    invDict[inversion] = pd.DataFrame({'partner_sample_id': metadata['sample_id'], \n",
    "                           'inversion':inversion, \n",
    "                           'mean_genotype': av_gts, \n",
    "                           'total_snp_tags':total_sites})\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = metadata[['sample_id', 'partner_sample_id', 'location2', 'species_gambiae_coluzzii']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gaardian_karyotypes = pd.concat(invDict.values()).rename(columns={'partner_sample_id':'sample_id'})\n",
    "gaardian_karyotypes = gaardian_karyotypes.merge(df).set_index('partner_sample_id').reset_index().drop(columns='sample_id')\n",
    "gaardian_karyotypes.to_csv(\"gaardian_karyotypes.tsv\", sep=\"\\t\") "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load karyo data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "gaardian_karyotypes = pd.read_csv(\"gaardian_karyotypes.tsv\", sep=\"\\t\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>partner_sample_id</th>\n",
       "      <th>inversion</th>\n",
       "      <th>mean_genotype</th>\n",
       "      <th>total_snp_tags</th>\n",
       "      <th>location2</th>\n",
       "      <th>species_gambiae_coluzzii</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WA-2001</td>\n",
       "      <td>2La</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>209</td>\n",
       "      <td>Adumanu</td>\n",
       "      <td>gambiae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WA-2001</td>\n",
       "      <td>2Rj</td>\n",
       "      <td>0.060606</td>\n",
       "      <td>99</td>\n",
       "      <td>Adumanu</td>\n",
       "      <td>gambiae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WA-2001</td>\n",
       "      <td>2Rb</td>\n",
       "      <td>0.123209</td>\n",
       "      <td>349</td>\n",
       "      <td>Adumanu</td>\n",
       "      <td>gambiae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WA-2001</td>\n",
       "      <td>2Rc_col</td>\n",
       "      <td>0.438596</td>\n",
       "      <td>57</td>\n",
       "      <td>Adumanu</td>\n",
       "      <td>gambiae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WA-2001</td>\n",
       "      <td>2Rc_gam</td>\n",
       "      <td>0.061224</td>\n",
       "      <td>49</td>\n",
       "      <td>Adumanu</td>\n",
       "      <td>gambiae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3390</th>\n",
       "      <td>WA-2500</td>\n",
       "      <td>2Rb</td>\n",
       "      <td>0.086207</td>\n",
       "      <td>348</td>\n",
       "      <td>Domenase</td>\n",
       "      <td>coluzzii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3391</th>\n",
       "      <td>WA-2500</td>\n",
       "      <td>2Rc_col</td>\n",
       "      <td>0.122807</td>\n",
       "      <td>57</td>\n",
       "      <td>Domenase</td>\n",
       "      <td>coluzzii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3392</th>\n",
       "      <td>WA-2500</td>\n",
       "      <td>2Rc_gam</td>\n",
       "      <td>0.020408</td>\n",
       "      <td>49</td>\n",
       "      <td>Domenase</td>\n",
       "      <td>coluzzii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3393</th>\n",
       "      <td>WA-2500</td>\n",
       "      <td>2Rd</td>\n",
       "      <td>0.081633</td>\n",
       "      <td>147</td>\n",
       "      <td>Domenase</td>\n",
       "      <td>coluzzii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3394</th>\n",
       "      <td>WA-2500</td>\n",
       "      <td>2Ru</td>\n",
       "      <td>0.033898</td>\n",
       "      <td>177</td>\n",
       "      <td>Domenase</td>\n",
       "      <td>coluzzii</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3395 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     partner_sample_id inversion  mean_genotype  total_snp_tags location2  \\\n",
       "0              WA-2001       2La       0.000000             209   Adumanu   \n",
       "1              WA-2001       2Rj       0.060606              99   Adumanu   \n",
       "2              WA-2001       2Rb       0.123209             349   Adumanu   \n",
       "3              WA-2001   2Rc_col       0.438596              57   Adumanu   \n",
       "4              WA-2001   2Rc_gam       0.061224              49   Adumanu   \n",
       "...                ...       ...            ...             ...       ...   \n",
       "3390           WA-2500       2Rb       0.086207             348  Domenase   \n",
       "3391           WA-2500   2Rc_col       0.122807              57  Domenase   \n",
       "3392           WA-2500   2Rc_gam       0.020408              49  Domenase   \n",
       "3393           WA-2500       2Rd       0.081633             147  Domenase   \n",
       "3394           WA-2500       2Ru       0.033898             177  Domenase   \n",
       "\n",
       "     species_gambiae_coluzzii  \n",
       "0                     gambiae  \n",
       "1                     gambiae  \n",
       "2                     gambiae  \n",
       "3                     gambiae  \n",
       "4                     gambiae  \n",
       "...                       ...  \n",
       "3390                 coluzzii  \n",
       "3391                 coluzzii  \n",
       "3392                 coluzzii  \n",
       "3393                 coluzzii  \n",
       "3394                 coluzzii  \n",
       "\n",
       "[3395 rows x 6 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gaardian_karyotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>partner_sample_id</th>\n",
       "      <th>inversion</th>\n",
       "      <th>mean_genotype</th>\n",
       "      <th>total_snp_tags</th>\n",
       "      <th>location2</th>\n",
       "      <th>species_gambiae_coluzzii</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>WA-2002</td>\n",
       "      <td>2La</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>209</td>\n",
       "      <td>Adumanu</td>\n",
       "      <td>coluzzii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>WA-2003</td>\n",
       "      <td>2La</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>209</td>\n",
       "      <td>Kente</td>\n",
       "      <td>coluzzii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>WA-2004</td>\n",
       "      <td>2La</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>209</td>\n",
       "      <td>Kente</td>\n",
       "      <td>coluzzii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>WA-2005</td>\n",
       "      <td>2La</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>209</td>\n",
       "      <td>Kente</td>\n",
       "      <td>coluzzii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>WA-2006</td>\n",
       "      <td>2La</td>\n",
       "      <td>0.009569</td>\n",
       "      <td>209</td>\n",
       "      <td>Odumto</td>\n",
       "      <td>coluzzii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3360</th>\n",
       "      <td>WA-2496</td>\n",
       "      <td>2La</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>209</td>\n",
       "      <td>Domenase</td>\n",
       "      <td>coluzzii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3367</th>\n",
       "      <td>WA-2497</td>\n",
       "      <td>2La</td>\n",
       "      <td>0.004785</td>\n",
       "      <td>209</td>\n",
       "      <td>Domenase</td>\n",
       "      <td>coluzzii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3374</th>\n",
       "      <td>WA-2498</td>\n",
       "      <td>2La</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>209</td>\n",
       "      <td>Domenase</td>\n",
       "      <td>coluzzii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3381</th>\n",
       "      <td>WA-2499</td>\n",
       "      <td>2La</td>\n",
       "      <td>0.995192</td>\n",
       "      <td>208</td>\n",
       "      <td>Domenase</td>\n",
       "      <td>coluzzii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3388</th>\n",
       "      <td>WA-2500</td>\n",
       "      <td>2La</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>209</td>\n",
       "      <td>Domenase</td>\n",
       "      <td>coluzzii</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>422 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     partner_sample_id inversion  mean_genotype  total_snp_tags location2  \\\n",
       "7              WA-2002       2La       0.000000             209   Adumanu   \n",
       "14             WA-2003       2La       0.000000             209     Kente   \n",
       "21             WA-2004       2La       1.000000             209     Kente   \n",
       "28             WA-2005       2La       1.000000             209     Kente   \n",
       "35             WA-2006       2La       0.009569             209    Odumto   \n",
       "...                ...       ...            ...             ...       ...   \n",
       "3360           WA-2496       2La       2.000000             209  Domenase   \n",
       "3367           WA-2497       2La       0.004785             209  Domenase   \n",
       "3374           WA-2498       2La       0.000000             209  Domenase   \n",
       "3381           WA-2499       2La       0.995192             208  Domenase   \n",
       "3388           WA-2500       2La       2.000000             209  Domenase   \n",
       "\n",
       "     species_gambiae_coluzzii  \n",
       "7                    coluzzii  \n",
       "14                   coluzzii  \n",
       "21                   coluzzii  \n",
       "28                   coluzzii  \n",
       "35                   coluzzii  \n",
       "...                       ...  \n",
       "3360                 coluzzii  \n",
       "3367                 coluzzii  \n",
       "3374                 coluzzii  \n",
       "3381                 coluzzii  \n",
       "3388                 coluzzii  \n",
       "\n",
       "[422 rows x 6 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "karyo = gaardian_karyotypes.loc[(gaardian_karyotypes['species_gambiae_coluzzii'] == 'coluzzii') & (gaardian_karyotypes['inversion'] == '2La')]\n",
    "karyo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_70/2162325607.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  karyo['karyotype']  = karyo.mean_genotype.round()\n"
     ]
    }
   ],
   "source": [
    "karyo['karyotype']  = karyo.mean_genotype.round()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "inversion  location2      karyotype\n",
       "2La        Adansi Apagya  0.0           9\n",
       "                          1.0           5\n",
       "                          2.0           1\n",
       "           Adansi-Krom    0.0           4\n",
       "                          1.0           2\n",
       "           Adumanu        0.0           4\n",
       "                          1.0           1\n",
       "           Anhwiaso       0.0           2\n",
       "           Ankaako        0.0           5\n",
       "                          1.0           5\n",
       "                          2.0           2\n",
       "           Annorkrom      1.0          25\n",
       "                          0.0          17\n",
       "                          2.0           4\n",
       "           Anwona         1.0           3\n",
       "                          0.0           2\n",
       "           Bogyawe        0.0           2\n",
       "                          1.0           1\n",
       "           Domenase       1.0          29\n",
       "                          0.0          15\n",
       "                          2.0          13\n",
       "           Dompoase       0.0           1\n",
       "           Fumso          0.0           1\n",
       "           Jacobu         0.0           1\n",
       "                          1.0           1\n",
       "           Kente          1.0          13\n",
       "                          0.0           9\n",
       "                          2.0           1\n",
       "           Kokotro        1.0          12\n",
       "                          0.0           9\n",
       "                          2.0           3\n",
       "           Koniyaw        0.0          18\n",
       "                          1.0          12\n",
       "                          2.0           1\n",
       "           Mprakyire      1.0           4\n",
       "                          2.0           3\n",
       "                          0.0           2\n",
       "           New Edubiase   1.0           9\n",
       "                          0.0           6\n",
       "           Nkotumso       1.0          15\n",
       "                          0.0          12\n",
       "                          2.0           3\n",
       "           Odumto         1.0          14\n",
       "                          0.0          13\n",
       "                          2.0           5\n",
       "           Pokukrom       1.0          22\n",
       "                          0.0          11\n",
       "                          2.0           4\n",
       "           Prisintease    0.0          13\n",
       "                          1.0           8\n",
       "                          2.0           3\n",
       "           Subin Camp     2.0           1\n",
       "           Wamase         0.0           7\n",
       "                          1.0           7\n",
       "                          2.0           2\n",
       "           Watreso        0.0          11\n",
       "                          1.0           6\n",
       "                          2.0           2\n",
       "           Yadome         0.0           3\n",
       "                          1.0           3\n",
       "Name: karyotype, dtype: int64"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "karyo.groupby(['inversion', 'location2']).karyotype.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.7.3 64-bit ('pysanj': conda)",
   "language": "python",
   "name": "python37364bitpysanjconda39958c5ba9ef41c3a5542f2bfe81dc48"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
